<!DOCTYPE html>
<html>

<head>
<link rel=stylesheet href=../style.css />
<link rel=icon href="https://s0.wp.com/wp-content/themes/vip/czi/images/build/favicon.ico" />

<style>
img.screenshot {
border: 2px solid Grey; 
margin: 6px;
padding: 6px;
}
</style>
</head>

<body>

<h1>ExpressionMatrix2 case study 1 (partial - work in progress)</h1>

<p>
In this case study I illustrate the functionality of 
<a href="../../index.html"><code>ExpressionMatrix2</code> software</a>
on expression data from <a id=paper href=https://doi.org/10.1016/j.celrep.2017.10.030>
Darmanis et al., 2017, "Single-Cell RNA-Seq Analysis of Infiltrating Neoplastic Cells at the Migrating Front of Human Glioblastoma", Cell Reports <i>21</i>, 1399-1410, October 31, 2017</a>. 
Special thanks to the paper's authors for providing early access to their data.

<p>
This is a small data set with under 4,000 cells. 
The <code>ExpressionMatrix2</code> software is designed to comfortably run
on much larger numbers of cells (it was actually run on a data set with over 1 million cells without any downsampling),
so the compute time required to run the case study is minimal,
and all phases of the run can be performed interactively.

<p>
This page contains detailed information sufficient to run the <code>ExpressionMatrix2</code> software on
these data. It contains the following sections:

<p>
<a href=#GettingStarted>Getting started</a>

<br>
<a href=#SettingUp>Setting up the run</a>

<br>
<a href=#Downloading>Downloading input files</a>

<br>
<a href=#Input>The input phase of the case study</a>

<br>
<a href=#Compute1>Finding pairs of similar cells</a>

<br>
<a href=#RunServer>A first look at the data</a>

<br>
<a href=#Clustering>Clustering</a>

<br>
<a href=#Conclusion>Conclusion</a>

<p>
Running this case study can serve as a useful introduction
to performing a run on your own data.



<br><br><h2 id=GettingStarted>Getting started</h2>
<p>
To begin, use <a href=../GettingStarted.html>these directions</a> to make sure you have access
to the software and an environment in which it can run.
When this step is completed, your environment variable <code>PYTHONPATH</code>
will be set to a directory containing a working version of <code>ExpressionMatrix2.so</code>
appropriate to your platform. 
To run this case study you will need to use <code>ExpressionMatrix2</code> release 0.3.0 or later.



<br><br><h2 id=SettingUp>Setting up the run</h2>

<p>Make a copy of directory <code>tests/CaseStudy1</code> from the 
<code>ExpressionMatrix2</code> release you are using.
To avoid possible errors or problems due to 
incompatibilities between releases,
make sure to use the same release from which 
your <code>ExpressionMatrix2.so</code> originates.
In the rest of this page, I will call this copy the 
<code>CaseStudy1</code> directory, and assume that all commands are
run from here.

At this point, your <code>CaseStudy1</code> directory
only contains some Python scripts and a <code>README</code> file.



<br><br><h2 id=Downloading>Downloading input files</h2>
<p>
The run requires the following input files, which can be downloaded from <a href="https://www.dropbox.com/sh/e8c2m7wi87z1c5h/AAAuJIhJZEik9JcylNUyh08fa?dl=0">here</a>:
<ul>
<li><code>GBM_raw_gene_counts.csv</code>
<li><code>GBM_metadata.csv</code>
</ul>
When your downloading is complete, move these files to your 
<code>CaseStudy1</code> directory. 
These files will be accessed with read-only access.
You may want to protect them by changing their access
permissions accordingly, for example using the following command:
<p>
<code>chmod ugo-w *.csv</code>

<p>
The expression matrix in file <code>GBM_raw_gene_counts.csv</code>
consists of raw read counts for each gene,
which are integer numbers.
However, the <code>ExpressionMatrix2</code> software
accepts as input expression counts which are floating point numbers,
for generality. This allows applying, if desired, pre-normalization steps
and/or non-linear transformations of the raw read counts
(for exampe, logarithmic transformations to emphasize the importance
of low expression genes).
   


<br><br><h2 id=Input>The input phase of the case study</h2>
<p>
You can begin the run using the following command:
<p>
<code>./input.py</code>
<p>
This causes the <code>ExpressionMatrix2</code>
software to read the input files and create a binary 
representation of the expression matrix 
in <code>GBM_raw_gene_counts.csv</code>
and of the cell meta data in <code>GBM_metadata.csv</code>.
This will run quickly (under one minute), 
and the run time will be dominated by the
time to read and parse input file <code>GBM_raw_gene_counts.csv</code>
containing the expression matrix.

<p> 
After this, the two input files will no longer be needed.
Instead, all required information is stored in binary files in the
<code>data</code> directory that was created in your
<code>CaseStudy1</code> directory.
Additional commands to be run later will create
additional binary files in the <code>data</code>
directory that will contain various types of
data structures used by the <code>ExpressionMatrix2</code> software.
These binary files provide fast sequential or random access to
the required data structures and are quite compact.
At this point the size of the <code>data</code> directory
is about half the size of the input csv files, and
is dominated by the space necessary to store the expression
matrix in a compact, sparse format
(file <code>data/CellExpressionCounts.data</code>).

<p>
Note that, if for any reason you want to restart the run,
you need to remove the <code>data</code>
directory manally before running <code>input.py</code> again.
This is a safety feature to prevent inadvertent 
data loss.

<p>
The scripts in the <code>CaseStudy1</code> directory
assume you are using a platform that uses Python 3.
If you use Python 2, you need to use the following instead of the
above command:
<p>
<code>python input.py</code>
<p>
This also applies to all invocations of Python scripts
for the rest of the case study.

<p>
See <a href=input.py.html>here</a> for details of <code>input.py</code>.



<br><br><h2 id=Compute1>Finding pairs of similar cells</h2>
<p>
With the expression data and cell meta data now stored
in the binary formats required by the <code>ExpressionMatrix2</code>
software,, the next step is to find pairs of cells that have similar 
gene expressions, with similarities computed taking all genes
into account. This is done using command

<pre>
./compute1.py
</pre>

The computation is fast (only a few seconds) as it is done
using Locality Sensitive Hashing (LSH). 
See <a href=compute1.py.html>here</a> for details of <code>compute1.py</code>.



<br><br><h2 id=RunServer>A first look at the data</h2>
<p>
At this point we are ready to take a first look at the data
using the http server functionality of the <code>ExpressionMatrix2</code>
software. 

<h3>Starting the http server</h3>
To do this, we use script <code>runServer.py</code>
to start the <code>ExpressionMatrix2</code> in 
http server mode:
<pre>
./runServer.py
</pre>
The code in <code>runServer.py</code> is very simple. It simply
creates the <code>ExpressionMatrix</code> object just like the
<code>compute1.py</code> script did, then invokes function 
<a href=../PythonApiReference/_build/html/index.html#ExpressionMatrix2.ExpressionMatrix.explore>explore</a>
to start the http server.
When the server starts, it writes a message containing the port number
on which it is listening for connections - usually port 17100.
The server will continue to run until you interrupt it using Ctrl-C
or by killing the process.  However, most of the time
it will be doing nothing - it will just be waiting
for http requests from browsers.

<h3>Connecting a browser to the server</h3>
<p>
With the server running, we can use a browser to explore the data.
The browser communicates  with the server using the http protocol,
and can run on the same machine as the server, or on a different machine,
as long as the two machines can communicate over a network or
via the Internet.
<p>
In the simplest mode of operations, the browser runs on the same machine
as the server. Start the browser and point it to this URL:
<pre>
http://localhost:17100
</pre>
(Make sure to change the port number from 17100 to whatever 
port number the server is listening on, which you found
out when starting the server).

<p>
If the browser is running on a different machine,
enter the same URL but replace localhost
with the name or ip address of the machine
where the server is running.
For example, the URL could look like this:
<pre>
http://server.company.com:17100
</pre>
or
<pre>
http://192.51.100.57:17100
</pre>
For this to work, however, it may be necessary to change the configuration
of firewall software or firmware on the server or on one of the routers
in between by "opening" TCP port 17100, and preferrably the range
17100-17110 to allow some flexibility, as the server
will try higher numbered ports if port 17100 is not available.

<p style='text-transform:uppercase;'>
Security warning: depending on your network and firewall configuration,
it may be possible for anybody with network access to the server
to look at your data and manipulate them. 
Use appropriate security precautions to prevent
unwanted access to the data. 
Your IT department will be able to provide security advice.

<p>
The  <code>ExpressionMatrix2</code> software was developed
mainly using the Firefox and Chrome browsers.
However most browsers will work, including browsers available 
on cell phones and other mobile devices.


<h3>Exploring the data in the browser</h3>

<p>
When you connect your browser to the server, 
the browser window will look like this:
<br><img class=screenshot src=Screenshot-Summary.png width=800><br>
The page summarizes the number of cells genes in the expression matrix
we are looking at, and then displays the README file in the 
<code>CaseStudy</code> directory. At the top of the page are
several buttons that can be clicked to explore the data set in a variety of ways.
For example, you can display information about genes, cells and cell meta data. 
You can also create and manipulate gene sets and cell sets, which are
subsets of the genes and cells in yiour expression matrix. 
Initially only the <code>AllGenes</code> and <code>AllCells</code> sets exists,
but you are free to create additional sets in a variety of ways.

<p>
As an example of using this functionality to explore the data, click on 
<code>Cell Sets</code> and then, in the table of cell sets, click
on the hyperlink for cell set <code>AllCells</code>.
This will display a table of all the cells in this cell set
(which of course includes all 3589 cells in our expression matrix),
and also gives you the ability to select meta data fields to be displayed.
For example, in the field <code>Select cell metadata to display:</code>,
select <code>Selection</code>, <code>Location</code>, and <code>Sample.name</code> 
(in most browsers
you will need to press the Ctrl key while selecting the second and third fields).
Then click on <code>Redisplay table</code> to obtain a table of
cells including the meta data fields you selected:

<br><img class=screenshot src=Screenshot-AllCells.png width=800><br>

<p>
You can then click on any cell to get detailed information about it:
<br><img class=screenshot src=Screenshot-CellTop.png width=800><br>

<p>
This also includes gene expression for the cell, which can be seen 
by scrolling down:
<br><img class=screenshot src=Screenshot-CellBottom.png width=800><br>

<p>
As another example of using the http server functionality to explore the data,
click on the <code>Cell meta data</code> button on the top, then
under <code>Histogram a meta data field</code> select the <code>AllCells</code>
cell set and select meta data name <code>Sample.name</code>,
then click on <code>Create histogram</code>.
<br><img class=screenshot src=Screenshot-Histogram.png width=800><br>

<p>
We can also create a contingency table showing the 
combined frequency of two meta data fields.
Click the <code>Cell meta data</code> button on the top, then
under <code>Contingency table for two meta data fields</code> select the <code>AllCells</code>
cell set and select meta data names <code>Sample.name</code> and <code>Location</code>,
then click on <code>Create contingency table</code>.
<br><img class=screenshot src=Screenshot-ContingencyTable.png width=800><br>

<p>
You can familiarize yourself with the rest of the functionality
to explore the data. Most of it should be self-explanatory.


<p>
We computed above pairs of similar cells
using script <code>compute1.py</code>.
We can now use these pairs of similar cells to create a
<i>cell similarity graph</i> - an undirected graph in which 
each vertex represents a cell, and two vertices
are joined by an edge if the corresponding edges have 
similarity greater than a chosen similarity threshold.
To avoid regions in the graph with very high connectivity,
we use a <code>k</code> 
parameter that causes only the 
best <code>k</code> neighbors to be kept for each vertex (k-Nearest
Neighbor or k-NN graph).

<p>
To create such a cell similarity graph (or simply <i>cell graph</i>)
you can click on button <code>Cell graphs</code> at the top of the browser window.
Then, you can select graph creation parameters:
<ul>
<li><code>Cell graph name</code>: for this graph we will use all of our 3589 cells,
and our similar cell pairs were computed taking all genes into account.
Therefore, we will choose <code>All</code> as the graph name.
<li><code>Cell set name</code>:
we pick the only choice available, <code>AllCells</code>, as we have not yet
created any additional cell sets.
<li><code>Similar pairs name</code>:
we pick the only choice available, <code>Lsh</code>, which is the only 
set of similar cell pairs we created. As a reminder,
when doing that we took into account all genes
when computing cell similarities.
<li><code>Similarity threshold</code>:
we will keep the default value of 0.5. Only cell pairs with similarity at least
0.5 will generate a graph edge. The similar cell pairs were computed
using a threshold of 0.2 (the default value for 
<code>similarPairs3</code>), so we will not lose
any edges due to insufficient information being stored in our similar cell pairs.
<li><code>Maximum connectivity</code>:
this is the <code>k</code> for the k-NN graph. 
We will keep the default value of 20.
The similar cell pairs were computed  using <code>k</code>=100
(the default value for 
<code>similarPairs3</code>), so we will not lose
any edges due to insufficient information being stored in our similar cell pairs.
</ul>
You can now click on <code>Create a new cell graph</code>.
This shows a page with a message stating the the graph had 84 isolated vertices, which were removed.
This means that there were 84 cells that no other cell with similarity at least 0.5.
These cells will not be present in the cell similarity graph.
Clicking on <code>Continue</code> gets you back to the <code>Cell graphs</code>
page, which now shows the cell graph you just created, named <code>All</code>.
You can display the graph by clicking on its name.
After a short wait to compute the graph layout using the <a href="http://graphviz.org">Graphviz</a> package,
the graph is displayed:
<br><img class=screenshot src=Screenshot-CellGraph-All-nocolor.png width=800><br>

The graph is displayed without edges for clarity and performance, but you
can turn the edges on, if desired, by unchecking the <code>Hide graph edges</code>
option and clicking on <code>Redraw graph</code>.

<p>
Note that the Graphviz package is used exclusively to compute the two-dimensional
layout of graphs for display. This layout is not used for any other
purpose, including clustering (more information on clustering follows).
It is just a visual aid.

<p>
You can experiment using the pink buttons to make the graphics larger, zoom in or out
to visualize specific regions of the graph, and
change the display size of vertices and the thickness of edges.

<p>
You can also color the graph in a variety of ways that should be self-explanatory.
This includes coloring by expression of a specific gene,
coloring by similarity to a specific cell, or coloring by
categorical or numerical meta data. 
You can also click on a vertex to display the page for the corresponding cell.
This complements the coloring options to allow you to get 
an idea of the characteristics of cells present
in different regions of the graph.

<p>
The graph shows clumps that correspond to groups of cells with similar expression vectors. 
In an overly simplified view of cell populations, each of these groups corresponds 
to a "cell type". The reality, of course, is much more complex. 
The <code>ExpressionMatrix2</code> software provides <i>clustering</i> functionality
which finds these clusters automatically, and I will cover this functionality below.
The <a href=#paper>Darmanis paper</a> used a commonly used clustering approach, 
<a href=https://en.wikipedia.org/wiki/T-distributed_stochastic_neighbor_embedding>t-SNE</a>, 
to classify the cells in groups, and each cell contains two meta data fields
that describe the clustering computed by t-SNE:
<ul>
<li>Meta data field <code>Cluster_2d</code> 
contains the id of the cluster each cell was assigned to.
<li>Meta data field <code>Cluster_2d_color</code> contains the color assigned to
the cluster each cell was assigned to.
These colors are the same colors used in Figure 2(A) of the <a href=#paper>Darmanis paper</a>.
The colors are stored as strings in HTML  
<a href='https://en.wikipedia.org/wiki/Web_colors#Hex_triplet'>hex triplet</a>
format, that, is in the form #RRGGBB, where RR, GG, BB are hexadecimal bytes
representing the red, green, and blue intensities.
</ul>

<p>
To see where the t-SNE clusters are located in our cell similarity graph, 
we can redisplay the graph, coloring each cell in the color given by its
<code>Cluster_2d_color</code>. 
To do this, in the form above the graph display check the <code>Color by meta data field</code> button,
and in the selections to the right of that button
select meta data field <code>Cluster_2d_color</code>
and ask that the field be interpreted as a <code>color</code>.
The graph display will no look like this:
<br><img class=screenshot src=Screenshot-CellGraph-All-tsne.png width=800><br>

<p>
We see that the clumps in the graph display correspond roughly to the t-SNE clusters.



<br><br><h2 id=GettingStarted>Limiting the gene set used to compute cell similarity</h2>
<p>
The <code>ExpressionMatrix2</code> graph was constructed 
taking into account expression counts for all genes, 
which means that the similarity of two cells is defined
as the regression coefficient between the expression counts
of the two cells, taking all genes into account.
In contrast, the <a href=#paper>Darmanis paper</a>
used only 500 genes selected based on the variability of their
expression.
<p>
In the <code>ExpressionMatrix2</code> software,
we can use <i>gene sets</i> to restrict the set of genes used.
Clicking on the <code>Gene sets</code> button at the top of the browser page
leads to a page that offers a variety of ways to create a new gene set,
including by pasting a set of gene names. 
This option could be used to define a gene set identical to the one
of 500 genes used in the <a href=#paper>Darmanis paper</a>.
Another option is to use the dialog near the bottom of the page,
<code>Create a gene set using gene information content</code>.
This creates a gene set based on an information content criterion
(e. g. a gene that is equally expressed in all cells has zero information content,
a gene expressed only in one of N cells has information content log<sub>2</sub>N bits).
We can select <code>AllGenes</code> as the starting gene set,
<code>AllCells</code> as the cell set used to compute information content,
request an information content of at least two bits,
and name the resulting gene set <code>HighInformationGenes</code>.
The same operation could also be performed in Python code 
using function 
<a href=../PythonApiReference/_build/html/index.html#ExpressionMatrix2.ExpressionMatrix.createGeneSetUsingInformationContent>createGeneSetUsingInformationContent</a>.

<p>
The resulting gene set has 22442 genes out of the 23368 total.
You can use the <code>Gene sets</code> page to look at the new gene set in detail.
To see what genes were removed, you could create a new gene set
consisting of the set difference betwene gene sets <code>AllGenes</code>
and <code>HighInformationGenes</code>.
You can also show the computed information content of each gene by using the dialog under
<code>Show gene information content</code>
near the bottom of the <code>Gene sets</code> page.

<p>
Now that we have a new gene set, we can redo the computation of the
pairs of similar cells, this time taking into account only the
gene set we just created. 
This is done by <code>compute2.py</code>, which is very similar to
<code>compute1.py</code>, except that it requests using the gene set we just created.
To run it, make sure ot first interrupt the http server (using Ctrl+C),
then run command 

<pre>
./compute2.py
</pre>

<p>
Like <code>compute1.py</code>, this takes just a few seconds for this small data set, 
after which you can restart the http server:

<pre>
./runServer.py
</pre>

<p>
In the browser, click on the <code>Cell graphs</code> button.
You will notice that cell graph <code>All</code> we previously created
disappeared. This happens because graphs (contrary to gene sets
and cell sets) are currently not persistent. 
In addition to recreating it, you can also create a new graph,
which you can name 
<code>AllHighInfo</code>,
which still uses cell set <code>AllCells</code>,
but now uses the similar cell pairs <code>LshHighInfo</code>
just computed by <code>compute2.py</code>.
With coloring by t-SNE color as before and displayed with edges hidden, the graph looks like this:
<br><img class=screenshot src=Screenshot-CellGraph-AllHighInfo.png width=800><br>

<p>
The clumps are generally more separated. However the table shows that 1596
vertices (cells) were isolated and were removed. This means that for each
of the 1596 removed cells there was no other cell with similarity
greater than 0.5, the default similarity threshold used to construct the graph.
This means that almost half of the cells were removed.
This is not surprising, because we discarded genes that were expressed in many cells,
and were therefore generally increasing the similarity between cells.

<p>
To obtain a more informative graph, 
we can create a new graph
in the same way, but choosing a similarity threshold of 0.2 instead of 0.5.
Call this new graph <code>AllHighInfo2</code>.
Now the number of isolated cells that were removed decreases to 22.

<p>
With coloring by t-SNE color as before, the new graph looks like this:
<br><img class=screenshot src=Screenshot-CellGraph-AllHighInfo2.png width=800><br>

<p>
Clearly, it can be useful to experiment with the graph creation parameters.
Different choices can reveal different aspects of the data set.

<p>
The http server offers various means to investigate the details of this graph - for example,
what type of cells are present in each of the clumps. 
To do this, you can use one or more of the following:
<ul>
<li>Click on a vertex of the graph to get to the page describing the corresponding cell.
From that, find the most expressed genes in that cell 
(in gene set <code>HighInformationGenes</code>), then redisplay the graph, coloring
by expression of those genes.
<li>Hover the mouse on a vertex of the graph to get the cell id (a number)
of the corresponding cell. Then color the graph by similarity to that cell.
<li>Color the graph using one of the meta data fields. 
For this data set, coloring by meta data fields
<code>Selection</code>,
<code>Location</code>, and
<code>Sample.name</code>,
is particularly informative.
</ul>

<p>
For example, hovering the mouse in the red region on the right
of the display shows that one of the cells there is cell 2235.
Clicking on it shows the page for cell 2235, which reveals that the
most expressed gene is <code>CCL3</code>. 
Here is the display we get if we color the graph by similarity to cell 2235:
<br><img class=screenshot src=Screenshot-CellGraph-AllHighInfo2-BySimilarity.png width=800><br>

And here is graph colored by expression of gene <code>CCL3</code>:
<br><img class=screenshot src=Screenshot-CellGraph-AllHighInfo2-ByCCL3.png width=800><br>

This functionality can be used to explore gene expressions in the various "clumps"
visible in the graph.


<br><br><h2 id=Clustering>Clustering</h2>
<p>
The <code>ExpressionMatrix2</code> software also provides clustering functionality
which can be used to automatically find clumps or clusters in a cell similarity graph.
It uses the <a href="https://en.wikipedia.org/wiki/Label_Propagation_Algorithm">
Label Propagation Algorithm</a> to assign vertices to clusters.
Additional clustering algorithms may also be added in the future.

<p>
After clusters are computed, <code>ExpressionMatrix2</code> also computes
similarities between the average expression vectors of each cluster,
and creates a <i>Cluster Graph</i> in which each vertex represents a cluster,
and two vertices are joined by an undirected edge if the similarity
between then average expression vectors of the corresponding clusters exceeds a
given threshold. 

<p>
To compute the clustering, click on the <code>Clustering</code>
button near the top of the browser window, 
then click on "Run clustering and create a new cluster graph".
A table appears, requesting various parameters for the clustering computation. 
Here, specify <code>AllHighInfo2</code> as the cell graph.
Leave all parameters at their default values except for setting
the minimum number of cells for each cluster to 50
and the similarity threshold for graph edges to 0.2
(the same value we used when creating the cell similarity graph).
Use the same name <code>AllHighInfo2</code>
as the name of the cluster graph to be created (cell graphs and cluster graphs
use different name spaces, so it's ok to use the same name). 
Finally, click on <code>Run clustering</code>.
A page showing log output for the clustering appears. 
Here, click on <code>Continue</code> to get back to ther <code>Cluster graphs</code>
page. This page will now contain a link to the newly created cluster graph 
<code>AllHighInfo2</code>. 
Click on that link to get a display of the cluster graph:
<br><img class=screenshot src=Screenshot-ClusterGraph.png width=800><br>

<p>
Here, each circle corresponds to a cluster. You can hover on the circle
to find out how many cells are in each cluster, and also to see the most
expressed genes in each cluster.
You can also get more detailed information on a cluster,
including the average gene expression in the cluster and the list of cells that belong to it,
by clicking on it. 
Each edge in the display is labeled by the similarity between
the average gene expressions of the two clusters it joins.

<p>
You can also compare the gene expression of two clusters or sets of clusters.
To do this, in the page that displays the cluster graph, click on 
<code>Compare gene expression between clusters</code>.
In the dialog that appears, specify a "first set" consisting of cluster 0 
(the cluster displayed in green) and a "second set" consisting of cluster 3
(the cluster displayed in magenta/fuchsia color).
According to the display of the cluster graph, the similarity 
of the average expressions of these two clusters is 0.51.
Then click on <code>Compare</code>.
You get the following display - a scatter plot of the average expression counts
in the second sets versus those in the first set.
<br><img class=screenshot src=Screenshot-ClusterExpression.png width=800><br>

Here, each point corresponds to a gene, and you can hover on a point to get the gene name.
This allows you to easily locate genes tha are over-expressed or under-expressed
in one of the two clusters relative to the other. 
For example, the rightmost point in the display corresponds to gene <code>CCL3</code>,
and from its position we can see that this gene is roughly 5 times more
expressed in the first set (cluster 0) then in the second set (cluster 3).

<p>
You can do the same
comparison for arbitrary sets of clusters, not just individual clusters.

<p>
To see how the cluster correspond to clumps in the cell graph,
we can store the result of the clustering operation in a new cell meta data field.
To do this, click on the button labeled <code>Store the cluster ids in this cluster graph
in cell meta data name</code> after specifying <code>Cluster</code> as the name
of the cell meta data to be used.
Now go back to the display of cell graph <code>AllHighInfo2</code>,
and color it using the meta data field we just created, <code>Cluster</code>,
interpreted as a category.
The display of the cell graph will now look like this:
<br><img class=screenshot src=Screenshot-CellGraph-ByCluster.png width=800><br>

The color in this display correspond to the colors in the display of the cluster
graph, which allows you to locate each of the computed clusters in the cell graph.
You can then see that the general connectivity of the cel graph
is consistent with the graph connectivity of the cluster graph.
For example, we see from the cell graph that the green and bright purple
clusters (clusters 0 and 4) are close to each other, 
and in fact the cluster graph shows a similarity 0.51 between those clusters.
You can also see that the the relative placement of clusters in the two graphs 
are similar, even though the two-dimensional layouts,
which are used for display purposes only, are not necessarily consistent.

<p>
The cluster graph therefore describes in a synthethic way the
clumps in the cell graph and their relative similarities.
Of course, several parameters were used to construct the cell similarity
graph and the cluster graph, and the actual clusters will depend on the choice 
of these parameters. This is to be expected, as the type and amount of 
desired clustering depends on what questions one wants to answer.
Therefore, changing the creation parameters for the cell and cluster graphs 
can be used in the process of exploring and understanding the data.

<p>
Despite the above considerations, it is interesting to
compare the results of this clustering with 
the clustering obtained via t-SNE in the
<a href=#paper>Darmanis paper</a>.
To do this, click on the button labeled
o<code>Cell meta data</code>, then, under 
<code>Contingency table for two meta data fields</code>,
select cell set name <code>AllCells</code>
and meta data names <code>Cluster</code>
(the clustering we just computed)
and <code>cluster_2d</code>
(the t-SNE clustering from the <a href=#paper>Darmanis paper</a>).
This displays a contingency table of the two clusterings -
that is, a table containing the number of cells
with each combination of the two clusterings.

<br><img class=screenshot src=Screenshot-ContingencyTable-Clustering.png width=800><br>
<p>
For example, this shows that cluster 0 in our clustering (900 cells) is split into clusters 
7 and 8 in the t-SNE clustering.
Conversely, t-SNE cluster 8 (1329 cells) is split between clusters 0, 3, 4, 7, and 10
in our clustering.
We see from our cluster graph display that, for example, the similarity
between the average expressions of clusters 0 and 3 is only 0.51,
so it probably makes sense to keep these are two separate clusters.
For other clusters, there is more agreement between the two clustering.
For example, t-SNE cluster 9 largely corresponds to our cluster 2,
and t-SNE cluster 2 largely corresponds to our cluster 11.
As discussed above, the relative merits of each clustering
can be dependent on the type of information we are trying to extract.
The <code>ExpressionMatrix2</code> software allows quick exploration 
of clustering alternatives, by making it fast and easy to generate
multiple cell similarity graphs and cluster graphs with different 
creation parameters.



<br><br><h2 id=Conclusion>Conclusion</h2>
<p>
I used this case study to illustrate most of the functionality of
the <code>ExpressionMatrix2</code> software currently available, but not all of it. 
Most notably, this case study did not illustrate
the functionality to define <i>cell sets</i> - arbitrary subsets of
cells. 
Once a cell set is defined, it can be used to create a cell
similarity graph, which will include only cells in the set.
This way you can restrict your investigation to a subset of cells of interest.
This can facilitate data interpretation by reducing distractions
due to cell populations unrelated to the topic of interest.
</ul>


</body>
</html>
