<!DOCTYPE html>
<html>

<head>
<link rel=stylesheet href=style.css>
</head>

<body>

<h1>Documentation for GitHub repository chanzuckerberg/ExpressionMatrix2</h1>

<h2>Introduction</h2>

<p>
This repository contains software for analysis of gene expression data from single-cell RNA sequencing developed at <a href="https://chanzuckerberg.com">Chan-Zuckerberg Initiative</a>.

<p>
This code is at a prototype, pre-alpha stage. It is hoped that it can be useful in its current form, but it is likely that it contains bugs or other errors that impact its ability to give meaningful results. In addition, a validation of this code has not been performed. This prototype code is released in open source form with an <a href=LICENSE.html>MIT license</a>. 

<h2>Functionality and approach</h2>

<p>
A single-cell RNA sequencing run creates an expression matrix that contains, for each cell, the number of reads that mapped to each gene, often referred to as the expression count for that cell and gene. The ExpressionMatrix2 code processes such an expression matrix in ways that permit various types of analysis and visualization. 

<p>
It is possible to interpret each expression count as a coordinate in a high-dimensional space in which each dimension corresponds to a gene. The expression matrix then maps each cell to a point in this high-dimensional space.
Most approaches to analyzing these expression matrices involve an initial step of <a href="https://en.wikipedia.org/wiki/Dimensionality_reduction">dimensionality reduction</a> which projects these points to a space of much lower dimensionality (often 2), which is then used for additional processing such as clustering, visualization, and othet types of analysis. With such approaches, it is critical that the dimensionality reduction step is done in a way that minimizes the loss of information incurred. Several approaches have been developed with this goal in mind.

<p>
In contrast, in the ExpressionMatrix2 code we do not use a dimensionality reduction step. All processing is done in a space of dimensionality equal to the number of genes, and two-dimensional representations are used exclusively for visualization, and not for analysis purposes.

<p>
To allow this, we use a "Cell Similarity Graph", an undirected graph in which each vertex represents a cell. Two vertices are joined by an undirected edge if the expression count vectors of the corresponding cells are sufficiently similar, according to some definition of similarity. Each edge is labeled with the similarity between the cells corresponding to the vertices it joins.  The Cell Similarity Graph then shows "communities" - highly connected regions that corresponds to groups of cells with similar gene expressions. We can then use a clustering algorithm to detect and characterize these communities. The clustering algorithm only takes into accound the graph connectivity, without any reference to the high-dimensional space defined by the gene expression counts.

<p>
For visualization purposes only, the ExpressionMatrix2 code visualizes the graph on a two-dimensional screen using <a href="http://graphviz.org">Graphviz</a>, a well known graph layout package. Graphviz uses a force-directed layour algorithm to compute two-dimensional coordinates suitable to display the graph, thus effectively performing a dimensionality reduction scheme. However, we do not use the computed two-dimensional coordinates for any purpose other than visualization. The graph "communities" are usually visible as "clumps" in the graph. However, note that if two "clumps" are shown on top of each other, it is not necessarily the case that they are highly connected to each other: the graph layout algorithm may need to show two unconnected communities on top of each other due to shortcomings of the two-dimensional representation of the graph.



<h2>Getting started</h2>

<p> 
Use the following direction to get quickly started and run ExpressionMatrix2 on a toy dataset distributed as part of the repository.

<ul>

<li>
Use a Linux machine running Ubuntu 16.04 (the current long term support release of Ubuntu). Other distributions using similar versions of the Linux kernel, such as Linux Mint 18, will also work.
    
<li>Make sure the following packages, all available from the standard Ubuntu repositories, are installed:
<ul>
<li>The Boost libraries (package libboost-all-dev).
<li>Python3 (package python3-all-dev).
<li>Graphviz (package graphviz).
</ul>
If one or more of these packages are not installed, you can install them using command "apt install packageName", where packageName is as indicated above. This requires root access.

<li>
Create an empty directory for the test you are going to run.

<li>
In GitHub, navigate to the bin directory of the chanzuckerberg/ExpressionMatrix2 repository and down to the library you want to use. There may be more than one version, each named ExpressionMatrix2.so and located in a directory named from the date the library was generated. Click on the selected library, then "Download". Copy it to the empty directory you created for the test. (If you prefer, you can instead clone the repository and get the library from you cloned copy. This would require running command "git clone https://github.com/chanzuckerberg/ExpressionMatrix2").

<li> In the same way, copy to your test directory files ExpressionMatrix.csv, MetaData.csv, run.py, and runServer.py from the GitHub repository under tests/ToyTest1. These files contain the expression matrix and cell meta data for the test.

<li>Run command "./run.py". This will read the csv files and create binary data structures stored in mapped files in a new directory named data.

<li>Run command "./runServer.py". This starts the ExpressionMatrix2 code in a mode where if behaves as an http server that can be used for interactive visualization and analysis. This command continues running until you interrupt it. 

<li>While the server is running, start a web browser on the same machine, and point it to the following URL: <a href="http://localhost:17100">http://localhost:17100</a>.

<li>Interactively explore the functionality offered by the server. You will need a more real test case to perform any interesting visualization or analysis, but if you made it here you are ready to run the ExpressionMatrix2 on real data.

</ul>
    
</body>
</html>
