<!DOCTYPE html>
<html>

<head>

<link rel=stylesheet href=style.css />
<link rel=icon href="https://s0.wp.com/wp-content/themes/vip/czi/images/build/favicon.ico" />
<style>
h3 {
    display: inline;
}
</style>

</head>

<body>

<h1>Reference information for the Python API</h1>

<p>This document contains reference information on the Python API
provided by <code>ExpressionMatrix2.so</code>. All names documented below
are in the namespace of the <code>Expression2</code> module
and become available after the module is imported.

<p>This document includes the following sections:
<ul>
<li><a href=#ErrorHandling>Error handling</a>
<li><a href=#Threading>Threading considerations</a>
<li><a href=#Pair>Pair classes</a>
<li><a href=#Container>Container classes</a>
<li><a href=#Constants>Constants</a>
<li><a href=#CellGraphVertexInfo>Class <code>CellGraphVertexInfo</code></a>
<li><a href=#ClusterGraphCreationParameters>Class <code>ClusterGraphCreationParameters</code></a>
<li><a href=#ExpressionMatrix>Class <code>ExpressionMatrix</code></a>
<ul>
    <li><a href=#Constructors>Constructors</a>
    <li><a href=#Sizes>Sizes</a>
    <li><a href=#Genes>Genes</a>
    <li><a href=#Cells>Cells</a>
    <li><a href=#CellMetaDataSection>Cell meta data</a>
    <li><a href=#ExpressionCounts>Cell expression counts</a>
    <li><a href=#GeneSets>Gene sets</a>
    <li><a href=#CellSets>Cell sets</a>
    <li><a href=#CellSimilarity>Cell similarity</a>
    <li><a href=#SimilarPairs>Pairs of similar cells</a>
    <li><a href=#CellGraphs>Cell graphs</a>
    <li><a href=#ClusterGraphs>Cluster graphs</a>
    <li><a href=#Miscellaneous>Miscellaneous</a>
</ul>
<li><a href=#ExpressionMatrixCreationParameters>Class <code>ExpressionMatrixCreationParameters</code></a>
<li><a href=#NormalizationMethod><code>NormalizationMethod</code></a>
<li><a href=#ServerParameters>Class <code>ServerParameters</code></a>
<li><a href=#Debugging>Debugging and testing functions</a>
</ul>


<br><br><h2 id=ErrorHandling>Error handling</h2>
Most errors are reported by raising a Python <code>Exception</code> 
containing an explanatory message. 
Some functions, as noted below in the documentation for individual functions,
return <code>True</code> if the operation succeeded, <code>False</code> otherwise.

<br><br><h2 id=Threading>Threading considerations</h2>
The underlying C++ code is currently written as single-threaded and
therefore it uses data structures that don't support concurrent execution.
If the calling Python code is multithreaded, it should use a mutex 
to serialize calls into each <code>ExpressionMatrix</code> object.
However, concurrent calls to separate <code>ExpressionMatrix</code> objects are permitted.

<br><br><h2 id=Pair>Pair classes</h2>
A pair class has two data members named <code>first</code> and <code>second</code>.
The name of each pair class ends with <code>Pair</code> and reflects the types of the two data members.
All pair classes have additional names that reflect
specific usages and improve code readability.



<br><br><table style='margin:auto'>
<tr><th>Pair class name<th>Type of <code>first</code><th>Type of <code>second</code>

<tr><td id=IntIntPair>
    <code>IntIntPair<br id=IntPair>IntPair</code>
    <td><code>Integer</code><td><code>Integer</code>

<tr><td id=UintUintPair>
    <code>UintUintPair
    <br id=UintPair>UintPair
    <br id=CellIdCellIdPair>CellIdCellIdPair
    <br id=CellIdPair>CellIdPair</code>
    <td><code>Non-negative integer</code><td><code>Non-negative integer</code>

<tr><td id=StringStringPair>
    <code>StringStringPair
    <br id=StringPair>StringPair
    <br id=NameValuePair>NameValuePair</code><td><code>String</code><td><code>String</code>

<tr><td id=UintFloatPair>
    <code>UintFloatPair
    <br id=GeneIdFloatPair>GeneIdFloatPair
    <br id=ExpressionCount>ExpressionCount</code><td><code>Non-negative integer</code><td><code>Float</code>

</table>



<br><br><h2 id=Container>Container classes</h2>
A container class behaves similarly to a Python list
where all elements are of the same type:
<ul>
<li>Brackets can be used to access container elements 
in exactly the same way as for Python lists.
<li>Function <code>len</code> can be called for a container and returns
the number of elements in it.
<li>The pattern <code>for x in v</code> can be used to iterate
over elements of a container.
<li><code>append</code>, <code>insert</code>, and <code>del</code> can be used to add or remove
elements like for Python lists. However, inserting or removing elements not at the end of
the container is computationally expensive. 
</ul>

The name of each container class ends with <code>List</code> and reflects the type
of the data it contains. Most containers types have additional names that reflect
specific usages and improve code readability.



<br><br><table style='margin:auto'>
<tr><th>Names of container type<th>Types of<br>contained elements

<tr><td id=IntList><code>IntList</code><td><code>Integer</code>
<tr><td id=UintList>
    <code>UintList
    <br id=GeneList>GeneIdList
    <br id=CellIdList>CellIdList</code>
<td><code>Non-negative integer</code>

<tr><td id=UintUintPairList>
    <code>UintUintPairList
    <br id=UintPairList>UintPairList
    <br id=CellIdCellIdPairList>CellIdCellIdPairList
    <br id=CellIdPairList>CellIdPairList</code>
<td><code>
    <a href=#UintUintPair>UintUintPair</a><br>
    <a href=#UintPair>UintPair</a><br>
    <a href=#CellIdCellIdPair>CellIdCellIdPair</a><br>
    <a href=#CellIdPair>CellIdPair</a></code>

<tr><td id=StringList><code>StringList</code><td><code>String</code>

<tr><td id=StringStringPairList>
    <code>StringStringPairList
    <br id=StringPairList>StringPairList
    <br id=NameValuePairList>NameValuePairList
    <br id=CellMetaData>CellMetaData</code>
<td><code>
    <a href=#StringStringPair>StringStringPair</a><br>
    <a href=#StringPair>StringPair</a><br>
    <a href=#NameValuePair>NameValuePair</a></code>

<tr><td id=StringStringPairListList>
    <code>StringStringPairListList
    <br id=StringPairListList>StringPairListList
    <br id=NameValuePairListList>NameValuePairListList
    <br id=CellMetaDataList>CellMetaDataList</code>
<td><code>
    <a href=#StringStringPairList>StringStringPairList</a><br>
    <a href=#StringPairList>StringPairList</a><br>
    <a href=#NameValuePairList>NameValuePairList</a><br>
    <a href=#CellMetaData>CellMetaData</a></code>

<tr><td id=FloatList>
    <code>FloatList
<td><code>Float

<tr><td id=DoubleList>
    <code>DoubleList
<td><code>Float

<tr><td id=UintFloatPairList>
    <code>UintFloatPairList
    <br id=GeneIdFloatPairList>GeneIdFloatPairList
    <br id=ExpressionCountList>ExpressionCountList
<td><code>
    <a href=#UintFloatPair>UintFloatPair</a><br>
    <a href=#GeneIdFloatPair>GeneIdFloatPair</a><br>
    <a href=#ExpressionCount>ExpressionCount</a><br>

<tr><td id=UintFloatPairListList>
    <code>UintFloatPairListList
    <br id=GeneIdFloatPairListList>GeneIdFloatPairListList
    <br id=ExpressionCountListList>ExpressionCountListList
<td><code>
    <a href=#UintFloatPairList>UintFloatPairList</a><br>
    <a href=#GeneIdFloatPairList>GeneIdFloatPairList</a><br>
    <a href=#ExpressionCountList>ExpressionCountList</a><br>

<tr><td id=CellGraphVertexInfoList><code>CellGraphVertexInfoList</code>
<td><code><a href=#CellGraphVertexInfo>CellGraphVertexInfo</a></code>

</table>



<br><br><h2 id=Constants>Constants</h2>
<p>The <code>ExpressionMatrix2</code> module defines the following constants:
<ul>
<li><code id=invalidGeneId><b>invalidGeneId</b></code>: an <code>integer</code> used to indicate a missing/invalid/undefined gene id.
<li><code id=invalidCellId><b>invalidCellId</b></code>: an <code>integer</code> used to indicate a missing/invalid/undefined cell id.
</ul>


<br><br><h2 id=CellGraphVertexInfo>Class <code>CellGraphVertexInfo</code></h2>
<p>This class is used to give the Python code access
to information about the vertices of a cell similarity graph.
An object of this type contains information for a single vertex.

<p>
<code><b>CellGraphVertexInfo</b>()
</code>
<br>Return value: <code>CellGraphVertexInfo</code>.
<br>This constructor creates an unitialized <code>CellGraphVertexInfo</code> object.

<p>
<code>CellGraphVertexInfo.<b>cellId</b>
</code>
<br>Type: <code>integer</code>
<br>This data member contains the cell id of the cell corresponding to the vertex.

<p>
<code>CellGraphVertexInfo.<b>x</b>()
</code>
<br>Return value: <code>float</code>
<br>Returns the x coordinate of the vertex in the two-dimensional layout of
the graph the vertex belongs to.

<p>
<code>CellGraphVertexInfo.<b>y</b>()
</code>
<br>Return value: <code>float</code>
<br>Returns the y coordinate of the vertex in the two-dimensional layout of
the graph the vertex belongs to.


<br><br><h2 id=ClusterGraphCreationParameters>Class <code>ClusterGraphCreationParameters</code></h2>
<p>This class describes parameters to be used when running the label propagation clustering algorithm to create a
new cluster graph from an existing cell graph.
<p>
<code><b>ClusterGraphCreationParameters</b>()
</code>
<br>Return value: <code>ClusterGraphCreationParameters</code>.
<br>The default constructor creates a <code>ClusterGraphCreationParameters</code>
initialized with the default values of the clustering parameters listed below.

<p>
<code>ClusterGraphCreationParameters.<b>stableIterationCount</b>
</code>
<br>Type: <code>integer</code>
<br>Default value: <code>3</code>
<br>This data member specifies the number of stable iterations (iterations
without any changes in the cluster labels) after which the label propagation
algorithm stops.

<p>
<code>ClusterGraphCreationParameters.<b>maxIterationCount</b>
</code>
<br>Type: <code>integer</code>
<br>Default value: <code>100</code>
<br>This data member specifies the maximum number of iterations that the label propagation
algorithm will run.

<p>
<code>ClusterGraphCreationParameters.<b>seed</b>
</code>
<br>Type: <code>integer</code>
<br>Default value: <code>231</code>
<br>This data member specifies an integer used as a random seed to initialize the
label propagation clustering algorithm.

<p>
<code>ClusterGraphCreationParameters.<b>minClusterSize</b>
</code>
<br>Type: <code>integer</code>
<br>Default value: <code>100</code>
<br>The minimum number of cells in a cluster. Smaller clusters are discarded.

<p>
<code>ClusterGraphCreationParameters.<b>maxConnectivity</b>
</code>
<br>Type: <code>integer</code>
<br>Default value: <code>3</code>
<br>The maximum connectivity (k of the k-NN graph) for the cluster graph.
A vertex can actually have more edges than this value, if the extra edges
are required by the opposite vertex.

<p>
<code>ClusterGraphCreationParameters.<b>similarityThreshold</b>
</code>
<br>Type: <code>integer</code>
<br>Default value: <code>0.5</code>
<br>The minimum similarity for an edge between two vertices to be created.
The edge is only created if the similarity between the average expression
vectors for the corresponding clusters exceeds this value.

<br><br><h2 id=ExpressionMatrix>Class <code>ExpressionMatrix</code></h2>
<p>This is the top level class in the <code>ExpressionMatrix2</code> module. 
Most high level functionality is provided by this class.
Binary data files for an instance of this class are stored in a single directory on disk.
They are accessed as memory mapped files.

<h3 id=Constructors>Constructors</h3>

<p>
<code><b>ExpressionMatrix</b>(directoryName, expressionMatrixCreationParameters)
<br>directoryName : string
<br>expressionMatrixCreationParameters: <a href=#ExpressionMatrixCreationParameters>ExpressionMatrixCreationParameters</a>
</code>
<br>Return value: <code>ExpressionMatrix</code>.
<br>This constructor creates a new (empty) <code>ExpressionMatrix</code> object in the specified directory.
The directory must not exists. 

<p>
<code><b>ExpressionMatrix</b>(directoryName)
<br>directoryName : string
</code>
<br>Return value: <code>ExpressionMatrix</code>.
<br>This constructor can be used to access an existing <code>ExpressionMatrix</code> 
object in the specified directory.
The directory must exist. 



<h3 id=Sizes>Sizes</h3>

<p>
<code>ExpressionMatrix.<b>geneCount</b>()
</code>
<br>Return value: the number of genes currently defined (<code>integer</code>).

<p>
<code>ExpressionMatrix.<b>cellCount</b>()
</code>
<br>Return value: the number of cells currently defined (<code>integer</code>).


<h3 id=Genes>Genes</h3>

<p>
<code>ExpressionMatrix.<b>geneName</b>(geneId)
<br>geneId: integer
</code>
<br>Return value: <code>string</code>
<br>Returns the name of the gene with the given <code>geneId</code>.

<p>
<code>ExpressionMatrix.<b>addGene</b>(geneName)
<br>geneName: string
</code>
<br>Return value: <code>integer</code>
<br>Adds a gene with the specified name. 
Returns <code>True</code> if the gene was added
and <code>False</code> if the gene already existed.



<h3 id=Cells>Cells</h3>

<p>
<code>ExpressionMatrix.<b>addCell</b>(jsonString)
<br>jsonString: string
</code>
<br>Return value: <code>integer</code>
<br>Adds a cell to the system. The cell expression counts and meta data 
are given in a JSON string. See 
<a href=PythonApi.html#addCell>here</a> for information on the expected
format of the JSON string. Returns the cell id of the cell that
was just added. Cell ids begin at zero and increment by one each time
a cell is added.

<p>
<code>ExpressionMatrix.<b>addCells</b>(expressionCountsFileName, expressionCountsFileSeparators, metaDataFileName, metaDataFileSeparators)
<br>expressionCountsFileName: string
<br>expressionCountsFileSeparators: string
<br>metaDataFileName: string
<br>metaDataFileSeparators: string
</code>
<br>Return value: <code>None</code>
<br>Adds cells to the system, reading expression counts and cell meta data
from files. See
<a href=PythonApi.html#addCells>here</a> for more information.

<p>
<code>ExpressionMatrix.<b>addCellsFromHdf5</b>(fileName, cellNamePrefix, cellMetaData, totalExpressionCountThreshold)
<br>fileName: string
<br>cellNamePrefix: string
<br>cellMetaData: <a href=#StringStringPairList>StringStringPairList</a>
<br>totalExpressionCountThreshold: float
</code>
<br>Return value: <code>None</code>
<br>Adds cells to the system, reading expression counts from a file
in HDF5 format as created by the 10X Genomics pipeline. See
<a href=PythonApi.html#addCellsFromHdf5>here</a> for more information.

<p id=addCellsFromBioHub1>
<code>ExpressionMatrix.<b>addCellsFromBioHub1</b>(expressionCountsFileName, initialMetaDataCount, finalMetaDataCount, plateMetaDataFileName)
<br>expressionCountsFileName: string
<br>initialMetaDataCount: integer
<br>finalMetaDataCount: integer
<br>plateMetaDataFileName: string
</code>
<br>Return value: <code>None</code>
<br>Add to the system cells created by the BioHub pipeline (July 2017 version, for Illumina data).
See <a href=PythonApi.html#addCellsFromBioHub1>here</a> for more information.

<p id=addCellsFromBioHub2>
<code>ExpressionMatrix.<b>addCellsFromBioHub2</b>(plateFileName, totalExpressionCountThreshold)
<br>plateFileName: string
<br>totalExpressionCountThreshold: integer
</code>
<br>Return value: <code>None</code>
<br>Add to the system cells created by the BioHub pipeline (September 2017 version, for 10X data).
See <a href=PythonApi.html#addCellsFromBioHub2>here</a> for more information.





<p>
<code>ExpressionMatrix.<b>cellIdFromString</b>(cellName)
<br>cellName: string
</code>
<br>Return value: <code>integer</code>
<br>Returns the cell id corresponding to a given <code>CellName</code>,
or <code>invalidCellId</code> if no cell with the given <code>CellName</code>
exists.



<h3 id=CellMetaDataSection>Cell meta data</h3>
<p>Each cell has a set of meta data which consists of (name, value) pairs. 
Each cell is permitted to have a different set of meta data names
from all other cells.

<p>
<code>ExpressionMatrix.<b>addCellMetaData</b>(fileName)
<br>fileName: string
</code>
<br>Return value: <code>None</code>
<br>Add cell meta data reading it from a csv file. Each line of the file contains cell meta data values
for one cell. Cell meta data names are in the first line. The first column of the file contain cell names.
All cell names must already be present. 

<p>
<code>ExpressionMatrix.<b>getCellMetaDataValue</b>(cellId, metaDataName)
<br>cellId: integer
<br>metaDataName: string
</code>
<br>Return value: <code>string</code>
<br>Returns the value of the specified meta data name for the given cell.
Returns an empty string if the cell does not have that meta data name.
The only meta data field that all cells are guaranteed to have is <code>CellName</code>.

<p>
<code>ExpressionMatrix.<b>getCellMetaData</b>(cellId)
<br>cellId: integer
</code>
<br>Return value: <code><a href=#CellMetaData>CellMetaData</a></code>
<br>Returns all meta data pairs (name, value) for a given cell.

<p>
<code>ExpressionMatrix.<b>getCellsMetaData</b>(cellIdList)
<br>cellIdList: <a href=#CellIdList>CellIdList</a>
</code>
<br>Return value: <code><a href=#CellMetaDataList>CellMetaDataList</a></code>
<br>Returns all meta data pairs (name, value) for a set of cells
specified by <code>cellIdList</code>. Each element in the returned 
<code><a href=#CellMetaDataList>CellMetaDataList</a></code>
corresponds to the cell id at the same position in <code>cellIdList</code>.


<h3 id=ExpressionCounts>Cell expression counts</h3>

<p>
<code>ExpressionMatrix.<b>getCellExpressionCount</b>(cellId, geneId)
<br>cellId: integer
<br>geneId: integer
</code>
<br>Return value: float</code>
<br>Returns the expression count for a given cell and gene.
The returned value can be zero.
This can be slow as it requires a binary search.

<p>
<code>ExpressionMatrix.<b>getCellExpressionCountFromGeneName</b>(cellId, geneName)
<br>cellId: integer
<br>geneName: string
</code>
<br>Return value: <code>float</code>
<br>Returns the expression count for a given cell and gene.
The returned value can be zero.
This can be slow as it requires a binary search.

<p>
<code>ExpressionMatrix.<b>getCellExpressionCounts</b>(cellId)
<br>cellId: integer
</code>
<br>Return value: <code><a href=#ExpressionCountList>ExpressionCountList</a></code>
<br>Returns the non-zero expression counts for a given cell.
The returned container contains pairs of gene ids and the
corresponding expression counts for the given cell.

<p>
<code>ExpressionMatrix.<b>getCellsExpressionCount</b>(cellIds, geneId)
<br>cellIds: <a href=#CellIdList>CellIdList</a>
<br>geneId: integer
</code>
<br>Return value: <code><a href=#FloatList>FloatList</a></code>
<br>Returns the expression counts for a set of cells and for a given gene.
Some or all of the returned values can be zero.
This can be slow as it requires a binary search.

<p>
<code>ExpressionMatrix.<b>getCellsExpressionCountFromGeneName</b>(cellIds, geneName)
<br>cellIds: <a href=#CellIdList>CellIdList</a>
<br>geneName: string
</code>
<br>Return value: <code><a href=#FloatList>FloatList</a></code>
<br>Returns the expression counts for a set of cells and for a given gene.
Some or all of the returned values can be zero.
This can be slow as it requires a binary search.

<p>
<code>ExpressionMatrix.<b>getCellsExpressionCounts</b>(cellIds)
<br>cellIds: <a href=#CellIdList>CellIdList</a>
</code>
<br>Return value: <code><a href=#ExpressionCountListList>ExpressionCountListList</a></code>
<br>Returns the non-zero expression counts for a given set of cells.
The returned container contains, for each cell, a list of pairs of gene ids and the
corresponding expression counts.



<h3 id=GeneSets>Gene sets</h3>

<p>Functionality to create and manipulate gene sets is provided and described in this section.
A gene set is a subset of all the genes known to the <code>ExpressionMatrix</code> object.
A gene set named <code>AllGenes</code>, and consisting of all the genes 
known to the known to the <code>ExpressionMatrix</code> object, is always automatically
created and maintained (that is, updated if new genes are added).

<p>
<code>ExpressionMatrix.<b>createGeneSetUsingInformationContent</b>(existingGeneSetName, cellSetName, normalizationMethod, geneInformationContentThreshold, newGeneSetName)
<br>existingGeneSetName: string
<br>cellSetName: string
<br>normalizationMethod: <a href=#NormalizationMethod>NormalizationMethod</a>
<br>geneInformationContentThreshold: float
<br>newGeneSetName: string
</code>
<br>Return value: <code>None</code>
<br>Creates a new gene set consisting of genes that pass an information content 
test. The information content, in bits, of each gene in <code>existingGeneSetName</code>
is computed using <code>cellSetName</code> and 
the specified <code>NormalizationMethod</code>.
The new gene set <code>newGeneSetName</code> is created, consisting
of genes whose computed information content (in bits) exceeds
the specified <code>geneInformationContentThreshold</code>.
The information content is roughly related to the variability of the
expression of that gene in the specified cell set. 
Genes that are widely expressed have a low information content.
Genes that are very selectively expressed have a high information content.
A gene that is equally expressed in all cells has zero information content.
A gene that is expressed only in one cell has <i>log<sub>2</sub>N</i>
bits of information content, where <i>N</i> is the number of cells
in the cell set used to compute information content.
A <code>geneInformationContentThreshold</code> of 2 bits is often effective
in filtering out widely expressed genes.

<p>
<code>ExpressionMatrix.<b>createGeneSetIntersection</b>(inputGeneSetNames, outputGeneSetName)
<br>inputGeneSetNames: string
<br>outputGeneSetName: string
</code>
<br>Return value: <code>Boolean</code>
<br>Creates a new gene set as the intersection of two or more gene sets.
The names of the input gene sets specified as the first argument
must be separated by commas.
Returns <code>False</code> if the operation failed because 
one of the input gene sets does not exist
or the output gene set already exists, <code>True</code> otherwise.

<p>
<code>ExpressionMatrix.<b>createGeneSetUnion</b>(inputGeneSetNames, outputGeneSetName)
<br>inputGeneSetNames: string
<br>outputGeneSetName: string
</code>
<br>Return value: <code>Boolean</code>
<br>Creates a new gene set as the union of two or more gene sets.
The names of the input gene sets specified as the first argument
must be separated by commas.
Returns <code>False</code> if the operation failed because 
one of the input gene sets does not exist
or the output gene set already exists, <code>True</code> otherwise.

<p>
<code>ExpressionMatrix.<b>createGeneSetDifference</b>(inputGeneSetName0, inputGeneSetName1, outputGeneSetName)
<br>inputGeneSetName0: string
<br>inputGeneSetName1: string
<br>outputGeneSetName: string
</code>
<br>Return value: <code>Boolean</code>
<br>Creates a new gene set as the difference between two gene sets.
The newly created gene set consists of all genes that are in 
<code>inputGeneSetName0</code> but not in <code>inputGeneSetName1</code>.
Returns <code>False</code> if the operation failed because 
one of the input gene sets does not exist
or the output gene set already exists, <code>True</code> otherwise.



<h3 id=CellSets>Cell sets</h3>

<p>Functionality to create and manipulate cell sets is provided and described in this section.
A cell set is a subset of all the cells known to the <code>ExpressionMatrix</code> object.
A cell set named <code>AllCells</code>, and consisting of all the cells 
known to the known to the <code>ExpressionMatrix</code> object, is always automatically
created and maintained (that is, updated if new cells are added).

<p>
<code id=getCellSetNames>ExpressionMatrix.<b>getCellSetNames</b>()
</code>
<br>Return value: <code><a href=#StringList>StringList</a></code>
<br>Return a list containing the names of all currently defined 
cell sets.

<p>
<code>ExpressionMatrix.<b>createCellSetUsingMetaData</b>(cellSetName, metaDataName, regularExpression)
<br>cellSetName: string
<br>metaDataName: string
<br>regularExpression: string
</code>
<br>Return value: <code>Boolean</code> (<code>True</code> if the operation succeeded, <code>False</code> otherwise).
<br>Creates a new cell set <code>cellSetName</code> consisting of all cells for which the
specified <code>metaDataName</code> matches the given 
<code>regularExpression</code>. 
As a simple application example, suppose all cells contain a meta data name
<code>Tissue</code> that gives the type of tissue from which the cell was obtained.
Calling <code>createCellSetUsingMetaData</code> with <code>metaDataName='Tissue'</code>
and <code>regularExpression='Brain'</code> will create a new cell set consisting of
the cells for which the value of the <code>Tissue</code> meta data field
is <code>Brain</code>. More complex regular expressions can be used to
select values that satisfy a variety of criteria. Functionality
to specify numeric ranges for meta data fields interpreted as numbers
is not current available but may become available at a future time.

<p>
<code>ExpressionMatrix.<b>createCellSetUsingNumericMetaDataGreaterThan</b>(cellSetName, metaDataName, lowerBound)
<br>cellSetName: string
<br>metaDataName: string
<br>lowerBound: float
</code>
<br>Return value: <code>None</code>.
<br>Creates a new cell set <code>cellSetName</code> consisting of all cells for which the
specified <code>metaDataName</code> is numeric and greater than <code>lowerBound</code>. 

<p>
<code>ExpressionMatrix.<b>createCellSetUsingNumericMetaDataLessThan</b>(cellSetName, metaDataName, upperBound)
<br>cellSetName: string
<br>metaDataName: string
<br>upperBound: float
</code>
<br>Return value: <code>None</code>.
<br>Creates a new cell set <code>cellSetName</code> consisting of all cells for which the
specified <code>metaDataName</code> is numeric and less than <code>upperBound</code>. 

<p>
<code>ExpressionMatrix.<b>createCellSetUsingNumericMetaDataBetween</b>(cellSetName, metaDataName, lowerBound, upperBound)
<br>cellSetName: string
<br>metaDataName: string
<br>lowerBound: float
<br>upperBound: float
</code>
<br>Return value: <code>None</code>.
<br>Creates a new cell set <code>cellSetName</code> consisting of all cells for which the
specified <code>metaDataName</code> is numeric, greater than <code>lowerBound</code>, and
less than <code>upperBound</code>. 

<p>
<code>ExpressionMatrix.<b>createCellSetIntersection</b>(inputCellSetNames, outputCellSetName)
<br>inputCellSetNames: string
<br>outputCellSetName: string
</code>
<br>Return value: <code>Boolean</code>
<br>Creates a new cell set as the intersection of two or more cell sets.
The names of the input cell sets specified as the first argument
must be separated by commas.
Returns <code>False</code> if the operation failed because 
one of the input cell sets does not exist
or the output cell set already exists, <code>True</code> otherwise.

<p>
<code>ExpressionMatrix.<b>createCellSetUnion</b>(inputCellSetNames, outputCellSetName)
<br>inputCellSetNames: string
<br>outputCellSetName: string
</code>
<br>Return value: <code>Boolean</code>
<br>Creates a new cell set as the union of two or more cell sets.
The names of the input cell sets specified as the first argument
must be separated by commas.
Returns <code>False</code> if the operation failed because 
one of the input cell sets does not exist
or the output cell set already exists, <code>True</code> otherwise.

<p>
<code>ExpressionMatrix.<b>createCellSetDifference</b>(inputCellSetName0, inputCellSetName1, outputCellSetName)
<br>inputCellSetName0: string
<br>inputCellSetName1: string
<br>outputCellSetName: string
</code>
<br>Return value: <code>Boolean</code>
<br>Creates a new cell set as the difference between two cell sets.
The newly created cell set consists of all cells that are in 
<code>inputCellSetName0</code> but not in <code>inputCellSetName1</code>.
Returns <code>False</code> if the operation failed because 
one of the input cell sets does not exist
or the output cell set already exists, <code>True</code> otherwise.


<p>
<code>ExpressionMatrix.<b>getCellSet</b>(cellSetName)
<br>cellSetName: string
</code>
<br>Return value: <a href=#CellIdList>CellIdList</a>
<br>Returns the cell ids of the cells in the cell set with the
specified name. If the cell set does not exists,
returns an ampty container.



<h3 id=CellSimilarity>Cell similarity</h3>

<p>
<code>ExpressionMatrix.<b>computeCellSimilarity</b>(cellId0, cellId1)
<br>cellId0: integer
<br>cellId1: integer
</code>
<br>Return value: <code>float</code>
<br>Returns the similarity between the expression vectors of two cells,
specified by their cell ids. The similarity is computed using all
the genes in the system. It is computed as the 
<a href='https://en.wikipedia.org/wiki/Pearson_correlation_coefficient'>Pearson correlation coefficient</a>
of the expression vectors of the two cells. The computed similarity is always between -1 and 1,
and it is rarely negative.

<p>
<code>ExpressionMatrix.<b>computeApproximateLshCellSimilarity</b>()
</code>
<br>Only intended to be used for testing.
See the source code in the <code>ExpressionMatrix2/src</code> directory for more information. 

<p>
<code>ExpressionMatrix.<b>writeLshSimilarityComparison</b>()
</code>
<br>Only intended to be used for testing.
See the source code in the <code>ExpressionMatrix2/src</code> directory for more information. 

<p>
<code>ExpressionMatrix.<b>writeLshSimilarityComparisonSlow</b>()
</code>
<br>Only intended to be used for testing.
See the source code in the <code>ExpressionMatrix2/src</code> directory for more information. 



<h3 id=SimilarPairs>Pairs of similar cells</h3>

<p>
<code id=findSimilarPairs0>ExpressionMatrix.<b>findSimilarPairs0</b>(geneSetName, cellSetName, similarPairsName, k, similarityThreshold)
<br>geneSetName: string
<br>cellSetName: string
<br>similarPairsName: string
<br>k: integer
<br>similarityThreshold: float
</code>
<br>Return value: <code>None</code>
<br>Creates and stores a new object <code>similarPairsName</code> 
containing pairs of similar cells from <code>cellSetName</code>.
The computation is done by directly looping over all possible cell pairs
with both cells in <code>cellSetName</code>
and performing for each pair an exact computation of the
cell similarity, taking into account only genes in <code>geneSetName</code>.
For each cell, only the best (most similar) <code>k</code> or fewer similar cells are stored,
among those that exceed the specified <code>similarityThreshold</code>.
This computational cost of this function grows with the square
of the number of cells in <code>cellSetName</code>. 
The required computing time will typically be a few minutes for a few thousand cells
or several hours for a few tens of thousands cells.
When the number of cells exceeds a few thousands, it is more practical
to perform an approximate computation using 
<a href=#findSimilarPairs1>findSimilarPairs1</a>.

<p>
<code id=findSimilarPairs1>ExpressionMatrix.<b>findSimilarPairs1</b>(geneSetName, cellSetName, similarPairsName, k, similarityThreshold, lshCount, seed)
<br>geneSetName: string
<br>cellSetName: string
<br>similarPairsName: string
<br>k: integer
<br>similarityThreshold: float
<br>lshCount: integer
<br>seed: integer
</code>
<br>Return value: <code>None</code>
<br>Like <a href=#findSimilarPairs0>findSimilarPairs0</a>, but uses Locality-Sensitive Hashing (LSH) with <code>lshCount</code>
LSH hashes to speed up the computation of the similar pairs. 
Like <a href=#findSimilarPairs0>findSimilarPairs0</a>, this still loops over all possible cell pairs
with both cells in <code>cellSetName</code>
and therefore the computational cost still grows asymptotically with the square of the number of cells.
However the computation is orders of magnitudes faster. 
The computation is approximate, and the error decreases as <code>lshCount</code> increases.
For the suggested value <code>lshCount=1024</code>, the standard deviation of 
the pair similarity computed in this way is 0.05 or less.
Improved LSH functionality that avoids the loop over cell pairs will be provided
by <a href=#findSimilarPairs2>findSimilarPairs2</a> but is not yet available.
More documentation for <code>findSimilarPairs1</code> will become available at a later time. 


<p>
<code id=findSimilarPairs2>ExpressionMatrix.<b>findSimilarPairs2</b>()
</code>
<br>Development of this functionality is in progress.
It will provide LSH-based computation of similar cell pairs
without looping over all possible pairs of cells.

<p>
<code>ExpressionMatrix.<b>writeSimilarPairs</b>()
</code>
<br>Only intended to be used for testing.
See the source code in the <code>ExpressionMatrix2/src</code> directory for more information. 

<p>
<code>ExpressionMatrix.<b>analyzeSimilarPairs</b>()
</code>
<br>Only intended to be used for testing.
See the source code in the <code>ExpressionMatrix2/src</code> directory for more information. 



<h3 id=CellGraphs>Cell graphs</h3>

<p>A <em>cell graph</em> or <em>cell similarity graph</em>
is an undirected graph in which each vertex represents a cell.
An edge between two vertices is created if the corresponding cells
have similarity that exceeds a chosen threshold. 
In areas of high connectivity, only the edges with the highest
connectivity for each cell are kept.

<p>
<code id=getCellGraphNames>ExpressionMatrix.<b>getCellGraphNames</b>()
</code>
<br>Return value: <code><a href=#StringList>StringList</a></code>
<br>Return a list containing the names of all currently defined 
cell similarity graphs.


<p>
<code id=createCellGraph>ExpressionMatrix.<b>createCellGraph</b>(cellGraphName, cellSetName, similarPairsName, similarityThreshold, k)
<br>cellGraphName: string
<br>cellSetName: string
<br>similarPairsName: string
<br>similarityThreshold: float
<br>k: integer
</code>
<br>Return value: <code>None</code>
<br>Create a new cell similarity graph with name given as the first argument.
The remaining arguments specify the parameters to be used to construct the
graph. <code>similarityThreshold</code> is the minimum cell similarity 
required for an edge to be created. <code>k</code> controls the maximum
connectivity of the graph.


<p>
<code id=computeCellGraphLayout>ExpressionMatrix.<b>computeCellGraphLayout</b>(cellGraphName)
<br>graphName: string
</code>
<br>Return value: <code>None</code>
<br>Computes the two-dimensional layout for the graph with the given name.
The graph must have been previously created with a call to 
<code><a href=#createCellGraph>createCellGraph</a></code>.
This function must be called once before the first call to 
<code>getCellGraphVertices</code> for the graph.

<p>
<code>ExpressionMatrix.<b>getCellGraphVertices</b>(cellGraphName)
<br>cellGraphName: string
</code>
<br>Return value: <code><a href=#CellGraphVertexInfoList>CellGraphVertexInfoList</a></code>
<br>Returns information about the vertices of the cell graph with the given name.
The graph must have been previously created with a call to
<code><a href=#createCellGraph>createCellGraph</a></code>
and its two-dimensional layout must have been computed with a call to 
<code><a href=#computeCellGraphLayout>computeCellGraphLayout</a></code>.

<p>
<code>ExpressionMatrix.<b>getCellGraphEdges</b>(cellGraphName)
<br>cellGraphName: string
</code>
<br>Return value: <code><a href=#CellIdPairList>CellIdPairList</a></code>
<br>Returns the cell ids corresponding to the two vertices 
of each edge of the cell graph with the given name.
The graph must have been previously created with a call to
<code><a href=#createCellGraph>createCellGraph</a></code>.


<h3 id=ClusterGraphs>Cluster graphs</h3>

<p>A <em>cluster graph</em> 
is an undirected graph in which each vertex represents a cluster found in a cell graph.
An edge between two vertices is created if the corresponding clusters
have average expression vectors that are sufficiently similar.
In areas of high connectivity, only the edges with the highest
connectivity for each cluster are kept.


<p>
<code id=createClusterGraph>ExpressionMatrix.<b>createClusterGraph</b>(cellGraphName, clusterGraphCreationParameters, clusterGraphName)
<br>cellGraphName: string
<br>clusterGraphCreationParameters: <a href=#ClusterGraphCreationParameters>ClusterGraphCreationParameters</a>
<br>clusterGraphName: string
</code>
<br>Return value: <code>None</code>
<br>Create a new cluster graph by running label propagation clustering on an existing cell graph.



<p>
<code id=getClusterGraphVertices>ExpressionMatrix.<b>getClusterGraphVertices</b>(clusterGraphName)
<br>clusterGraphName: string
</code>
<br>Return value: <code><a href=#IntList>IntList</a></code>
<br>Return a list of the cluster ids for the vertices of an existing cluster graph.
Each vertex of the cluster graph corresponds to a cluster.



<p>
<code id=getClusterGraphGenes>ExpressionMatrix.<b>getClusterGraphGenes</b>(clusterGraphName)
<br>clusterGraphName: string
</code>
<br>Return value: <code><a href=#GeneList>GeneList</a></code>
<br>Return a list of the ids of the genes used to create a cluster graph.



<p>
<code id=getClusterCells>ExpressionMatrix.<b>getClusterCells</b>(clusterGraphName, clusterId)
<br>clusterGraphName: string
<br>clusterId: integer
</code>
<br>Return value: <code><a href=#CellIdList>CellIdList</a></code>
<br>Return a list of the cell ids ids for the cells a cluster.
The cluster is specified using the name of the cluster graph it belongs to, 
and its cluster id within that cluster graph. 



<p>
<code id=getClusterAverageExpression>ExpressionMatrix.<b>getClusterAverageExpression</b>(clusterGraphName, clusterId)
<br>clusterGraphName: string
<br>clusterId: integer
</code>
<br>Return value: <code><a href=#DoubleList>DoubleList</a></code>
<br>Return the average, L2-normalized, gene expression of a cluster.
The cluster is specified using the name of the cluster graph it belongs to, 
and its cluster id within that cluster graph. 
The values returned correspond one on one to the gene ids returned
by <a href=#getClusterGraphGenes>getClusterGraphGenes</a> for the same cluster graph.


<p>
The example code below illustrates some cluster graph functionality.
It assumes that the entire <code>ExpressionMatrix2</code>
module was imported in the global Python namespace.

<pre>
# Access a previously created ExpressionMatrix object.
e = ExpressionMatrix('data')

# Create a cell graph.
e.createCellGraph('AllCells', 'AllCells', 'Exact', 0.5, 20)

# Run clustering on the cell graph we just created. This creates a cluster graph.
e.createClusterGraph('AllCells', ClusterGraphCreationParameters(), 'AllCells')

# Get the genes used by this cluster graph.
genes = e.getClusterGraphGenes('AllCells')

# Get average expression for cluster 0.
counts = e.getClusterAverageExpression('AllCells', 0)

# Write out the gene expression.
for i in range(len(genes)):
    globalGeneId = genes[i]
    geneName = e.geneName(globalGeneId)
    print(i, globalGeneId, geneName, counts[i])
</pre>
<br>



<h3 id=Miscellaneous>Miscellaneous</h3>

<p>
<code>ExpressionMatrix.<b>explore</b>(serverParameters)
<br>serverParameters: <a href=#ServerParameters>ServerParameters</a>
</code>
<br>Return value: <code>None</code>
<br>Starts an http server that can be used, in conjunction with a Web browser,
to interact with the <code>ExpressionMatrix</code> object.

<p>
<code>ExpressionMatrix.<b>testExpressionMatrixSubset</b>(cellId0, cellId1)
<br>cellId0: integer
<br>cellId1: integer
</code>
<br>Return value: <code>None</code>
<br>Only intended to be used
to debug or test the <code>ExpressionMatrix2</code> module.



<br><br><h2 id=ExpressionMatrixCreationParameters>Class <code>ExpressionMatrixCreationParameters</code></h2>
<p>This class is used to store creation parameters for a new <code>ExpressionMatrix</code>
object. 

<p>
<code><b>ExpressionMatrixCreationParameters</b>()
</code>
<br>Return value: <code>ExpressionMatrixCreationParameters</code>.
<br>This constructor creates a <code>ExpressionMatrixCreationParameters</code> object 
filled in with default values.

<p>
<code>ExpressionMatrixCreationParameters.<b>geneCapacity</b></code>
<br>Type: <code>integer</code>.
<br>Controls the size of the hash table used to hold genes.
For performance, the number of genes should not go above half of the value
specified here. Rehashing is currently not supported, so this
size cannot be changed after creation. 

<p>
<code>ExpressionMatrixCreationParameters.<b>cellCapacity</b></code>
<br>Type: <code>integer</code>.
<br>Controls the size of the hash table used to hold cells.
For performance, the number of cells should not go above half of the value
specified here. Rehashing is currently not supported, so this
size cannot be changed after creation. 

<p>
<code>ExpressionMatrixCreationParameters.<b>cellMetaDataNameCapacity</b></code>
<br>Type: <code>integer</code>.
<br>Controls the size of the hash table used to hold
cell meta data names.
For performance, the number of distinct cell meta data names
should not go above half of the value
specified here. Rehashing is currently not supported, so this
size cannot be changed after creation. 

<p>
<code>ExpressionMatrixCreationParameters.<b>cellMetaDataValueCapacity</b></code>
<br>Type: <code>integer</code>.
<br>Controls the size of the hash table used to hold
cell meta data values.
For performance, the number of distinct cell meta data values 
should not go above half of the value
specified here. Rehashing is currently not supported, so this
size cannot be changed after creation. 


<br><br><h2 id=NormalizationMethod><code>NormalizationMethod</code></h2>
<p>This is an enumerated type that defines the normalization method
to be used for cell expression vectors. 

<p>
<code>NormalizationMethod.<b>none</b></code>
<br>This specifies that no normalization should be applied to 
cell expression vectors. Cell expression vectors are returned or
used directly as specified when each cell was entered into the system.

<p>
<code>NormalizationMethod.<b>L1</b></code>
<br>This requests L1 normalization for 
cell expression vectors. This means that each cell expression vector
is scaled in such a way that the sum of its element is 1
before being returned or used.
With this normalization, each element of the cell expression vector
represents the fraction in each gene of the total
expression count for the cell.
L1 normalization is sensitive to the gene set in use - that is,
removing or adding genes changes the values of 
all entries in the cell expression vector.

<p>
<code>NormalizationMethod.<b>L2</b></code>
<br>This requests L2 normalization for 
cell expression vectors. This means that each cell expression vector
is scaled in such a way that the sum of the squares of its element is 1
before being returned or used.
In other words, expression counts are interpreted as components
of a vector in Euclidean space, and the returned vector is a unit
vector in that space.
L2 normalization is sensitive to the gene set in use - that is,
removing or adding genes changes the values of 
all entries in the cell expression vector.



<br><br><h2 id=ServerParameters>Class <code>ServerParameters</code></h2>
<p>This class specifies parameters for an http server to be
started by calling <code>ExpressionMatrix.explore</code>.

<p>
<code><b>ServerParameters</b>()
</code>
<br>Return value: <code>ServerParameters</code>.
<br>This constructor creates a <code>ServerParameters</code> object 
filled in with default values.

<p>
<code>ServerParameters.<b>port</b></code>
<br>Type: <code>integer</code>.
<br>The port number on which the server will be listening for connections. 
Its default value is 17100. On Linux, root access or other
security privileges are required to be able to listen on a port number less than 1024,
and many values velow 1024 are used for standard services.
If port 80 is specified, the user does not need to specify a port number to access the server,
becausde port 80 is the default port for the HTTP protocol.
The maximum allowed port number is 65535. 

<p>
<code>ServerParameters.<b>docDirectory</b></code>
<br>Type: <code>integer</code>.
<br>The path to the doc directory containing the documentation 
to be displayed by the server when the user clicks on the <i>Help</i> button.
This should point to a local copy of the <code>ExpressionMatrix2/doc</code> 
in the GitHub repository. If no value is specified (the default),
the server will still run but the <i>Help</i> button will not be available.



<br><br><h2 id=Debugging>Debugging and testing functions</h2>
<p>This section lists functions that are only intended to be used
to debug or test the <code>ExpressionMatrix2</code> module.
For more information about them, see the source code
in the <code>ExpressionMatrix2/src</code> directory.

<br>
<code>
<br>ExpressionMatrix2.<b>testMemoryMappedVector</b>()
<br>ExpressionMatrix2.<b>testMemoryMappedVectorOfLists</b>()
<br>ExpressionMatrix2.<b>testMemoryMappedStringTable</b>()
</code>



</body>
</html>
