<!DOCTYPE html>
<html>

<head>
<link rel=stylesheet href=style.css />
<link rel=icon href="https://s0.wp.com/wp-content/themes/vip/czi/images/build/favicon.ico" />
</head>

<body>

<h1>Running with real data</h1>

<p>
This assumes that you completed successfuly the test described under <a href=GettingStarted.html>Getting Started</a>. You can begin by copying the run.py and runServer.py scripts from that run to the directory you want to use for your run with real data. Those scripts will probably require some customization as described below, but are a useful starting point.



<h2>Input files</h2>

<p> 
The ExpressionMatrix2 code currently accepts input in three formats:

<ul>
<li>Comma separated format (csv) or similar delimited formats.
<li>HDF5 format as created by the 10x Genomics pipeline.
<li>A simple JSON format.
<li>Data created by various versions of the BioHub RNA sequencing pipeline.
</ul>

<p>
See <a href="PythonApi.html#addCells">here</a> for detailed information on how to use each of these three formats.



<h2>Expression matrix creation parameters</h2>

<p>
You will see several parameters in run.py that control creation of the expression matrix:

<ul>
<li>geneCapacity controls the maximum number of genes.
<li>cellCapacity controls the maximum number of cells.
<li>cellMetaDataNameCapacity controls the maximum number of distinct cell meta data names.
<li>cellMetaDataValueCapacity controls the maximum number of distinct cell meta data names.
</ul>

<p>
For good performance of various hash tables used to store the expression matrix, it is best if the actual number of each of these items stays below half the specified capacity. The current code does not allow rehashing (that is, changing table capacity after initialization), but future versions might. This will be particularly important if cells are to be added incrementally to the system.

<p>
When using the http server, the Summary page shows details of the capacity and actual number of these items.



<h2>Finding similar cell pairs</h2>

<p>
Before a graph can be created in the http server, it is necessary to compute at least a set of pairs of similar cells. This is done in run.py with a call to findSimilarPairs0, which uses the following parameters:

<ul>
<li>The names of the gene set and cell sets to be used for the computation.
<li>The name to be given to the set of similar pairs being computed. This name can then be specified to the http server when creating a cell similarity graph.
<li>The maximum number of similar cells that will be stored for each cell.
<li>The similarity threshold for a pair of similar pairs to be stored.
</ul>

<p>
For a large run, approximate computation of similar pairs is much faster than exact computation and gives very similar results. In both cases, the computing cost is O(N<sup>2</sup>), that is, proportional to the square of the number of cells. For a run with a few thousand cellsm exact computatioon will require hours, but approximate computation will require just minutes. It is possible to compute both exact and approximate similar pairs in the same run, as long as they are given different names in the call to findSimilarPairs0.

<p>
Future versions of the code will offer faster ways to find similar cell pairs that will have a better scaling than O(N<sup>2</sup>).


<h2>Binary Compatibility warning</h2>

<p> 
The ExpressionMatrix2 code uses binary files mapped in memory to store its data structures. It is likely that the binary format of these files will change as the code gets developed. This means that newer versions of ExpressionMatrix2 will not be able to access binary files created by older versions. In other words, <span style='text-transform:uppercase'>binary compatibility between versions is not guaranteed</span>. For this reason, <span style='text-transform:uppercase'>the binary files should not be used for long-term storage of expression matrix data</span>.

</body>
</html>
