<!DOCTYPE html>
<html>

<head>
<link rel=stylesheet href=style.css />
<link rel=icon href="https://s0.wp.com/wp-content/themes/vip/czi/images/build/favicon.ico" />
</head>

<body>

<h1>Running with real data</h1>

<p>
This assumes that you completed successfuly the test described under <a href=GettingStarted.html>Getting Started</a>. You can begin by copying the run.py and runServer.py scripts from that run to the directory you want to use for your run with real data. Those scripts will probably require some customization as described below, but are a useful starting point.



<h2>Input files</h2>

<p>
You will need an input text file containing the expression matrix. This file can be a standard comma separated file (csv), which follows the following rules:

<ul>
<li>Fields in each row are separated by commas.
<li>A field cannot contain a comma, unless the entire field is quoted or the comma is escaped (preceeded by an extra "\" character). 
<li>If a field contains quotes, the quotes in the field must be escaped (preceeded by an extra "\" character).
</ul>

<p>
The expression matrix file has a row for each gene and a column for each cell, plus an initial row containing cell names and an initial column containing gene names. Note the first field of the first row is ignored (can be anything), but must be present, so it must be manually added if it is not present. See the expression matrix file used in <a href=GettingStarted.html>Getting Started</a> for an example. Except for the initial row and column, each entry in the file contains a number which is the expression count for the gene and cell corresponding to the row and column that the field belongs to. These expression counts do not have to be integers (floating-point values are accepted), so prenormalization of the data can be done, if desired (for example to convert reads to reads per Megabase).

<p>
You need to modify the call to addCells in run.py so the first argument is the name of your file containing the expression matrix. The second argument specifies the separator used in the file. If your file is a comma separated file, you can leave it as is. Otherwise, you can replace it with the appropriate separator.

<p>
Some pipelines create expression matrices that contain rows, near the beginning or the end of the file, which contain extraneous information other than gene expression data. Such information should be removed. Also, expression counts for artificial ERCC sequence should be removed from, the expression file, if present.

<p>
An input file containing cell meta data can also be provided, but is not mandatory. The name of this file is the third argument toaddCells in run.py. It can be specified as an empty string if the file is not available. The cell meta data file contains a row for each cell and a column for each meta data field, plus an initial column containing cell names and an initial row containing meta data field names. See the meta data file used in <a href=GettingStarted.html>Getting Started</a> for an example. The fourth argument to addCells is the separator for the meta data file, and can be left as a comma if the meta data file is comma separated.

<p>
If either the expression file or the meta data file were generated by a Windows system, you will need to use the dos2unix command to convert the line ends. If dos2unix is not available, install it using command "apt install dos2unix" (this requires root privileges).


<h2>Expression matrix creation parameters</h2>

<p>
You will see several parameters in run.py that control creation of the expression matrix:

<ul>
<li>geneCapacity controls the maximum number of genes.
<li>cellCapacity controls the maximum number of cells.
<li>cellMetaDataNameCapacity controls the maximum number of distinct cell meta data names.
<li>cellMetaDataValueCapacity controls the maximum number of distinct cell meta data names.
</ul>

<p>
For good performance of various hash tables used to store the expression matrix, it is best if the actual number of each of these items stays below half the specified capacity. The current code does not allow rehashing (that is, changing table capacity after initialization), but future versions might. This will be particularly important if cells are to be added incrementally to the system.

<p>
When using the http server, the Summary page shows details of the capacity and actual number of these items.



<h2>Finding similar cell pairs</h2>

<p>
Before a graph can be created in the http server, it is necessary to compute at least a set of pairs of similar cells. This is done in run.py with a call to findSimilarPairs0, which uses the following parameters:

<ul>
<li>The name to be given to the set of similar pairs being computed. This name can then be specified to the http server when creating a cell similarity graph.
<li>The maximum number of similar cells that will be stored for each cell.
<li>The similarity threshold for a pair of similar pairs to be stored.
<li> A Boolean flag that is True if an exact computation of cell similarities is desired, False for an approximate computation. In the latter case, computations of cell similarities are performed using only the top genes for each cell, in number maxTermCountForApproximateSimilarityComputation specified as the last argument to addCells.
</ul>

<p>
For a large run, approximate computation of similar pairs is much faster than exact computation and gives very similar results. In both cases, the computing cost is O(N<sup>2</sup>), that is, proportional to the square of the number of cells. For a run with a few thousand cellsm exact computatioon will require hours, but approximate computation will require just minutes. It is possible to compute both exact and approximate similar pairs in the same run, as long as they are given different names in the call to findSimilarPairs0.

<p>
Future versions of the code will offer faster ways to find similar cell pairs that will have a better scaling than O(N<sup>2</sup>).


<h2>Binary Compatibility warning</h2>

<p> 
The ExpressionMatrix2 code uses binary files mapped in memory to store its data structures. It is likely that the binary format of these files will change as the code gets developed. This means that newer versions of ExpressionMatrix2 will not be able to access binary files created by older versions. In other words, <span style='text-transform:uppercase'>binary compatibility between versions is not guaranteed</span>. For this reason, <span style='text-transform:uppercase'>the binary files should not be used for long-term storage of expression matrix data</span>.

</body>
</html>
