<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>ExpressionMatrix2 Python API &mdash; ExpressionMatrix2  documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="ExpressionMatrix2  documentation" href="#" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">ExpressionMatrix2  documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="expressionmatrix2-python-api">
<h1>ExpressionMatrix2 Python API</h1>
<div class="toctree-wrapper compound">
<ul class="simple">
</ul>
</div>
<span class="target" id="module-ExpressionMatrix2"></span><p>Software for analysis, visualization, and clustering of gene expression data from single-cell RNA sequencing.</p>
<p>This page contains reference information for the Python API.
See <a class="reference external" href="../../../index.html">here</a> for the complete documentation.</p>
<div class="section" id="cellgraphvertexinfo">
<h2>CellGraphVertexInfo</h2>
<dl class="class">
<dt id="ExpressionMatrix2.CellGraphVertexInfo">
<em class="property">class </em><code class="descclassname">ExpressionMatrix2.</code><code class="descname">CellGraphVertexInfo</code></dt>
<dd><p>Class used to store information about a vertex of a cell graph.</p>
<dl class="method">
<dt id="ExpressionMatrix2.CellGraphVertexInfo.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.CellGraphVertexInfo</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="ExpressionMatrix2.CellGraphVertexInfo.cellId">
<code class="descname">cellId</code></dt>
<dd></dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.CellGraphVertexInfo.x">
<code class="descname">x</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.CellGraphVertexInfo</em><span class="sig-paren">)</span> &rarr; float</dt>
<dd></dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.CellGraphVertexInfo.y">
<code class="descname">y</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.CellGraphVertexInfo</em><span class="sig-paren">)</span> &rarr; float</dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="clustergraphcreationparameters">
<h2>ClusterGraphCreationParameters</h2>
<dl class="class">
<dt id="ExpressionMatrix2.ClusterGraphCreationParameters">
<em class="property">class </em><code class="descclassname">ExpressionMatrix2.</code><code class="descname">ClusterGraphCreationParameters</code></dt>
<dd><dl class="method">
<dt id="ExpressionMatrix2.ClusterGraphCreationParameters.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ClusterGraphCreationParameters</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="ExpressionMatrix2.ClusterGraphCreationParameters.stableIterationCount">
<code class="descname">stableIterationCount</code></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="ExpressionMatrix2.ClusterGraphCreationParameters.maxIterationCount">
<code class="descname">maxIterationCount</code></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="ExpressionMatrix2.ClusterGraphCreationParameters.seed">
<code class="descname">seed</code></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="ExpressionMatrix2.ClusterGraphCreationParameters.minClusterSize">
<code class="descname">minClusterSize</code></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="ExpressionMatrix2.ClusterGraphCreationParameters.maxConnectivity">
<code class="descname">maxConnectivity</code></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="ExpressionMatrix2.ClusterGraphCreationParameters.similarityThreshold">
<code class="descname">similarityThreshold</code></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="expressionmatrix">
<h2>ExpressionMatrix</h2>
<dl class="class">
<dt id="ExpressionMatrix2.ExpressionMatrix">
<em class="property">class </em><code class="descclassname">ExpressionMatrix2.</code><code class="descname">ExpressionMatrix</code></dt>
<dd><p>Top level object used to store and manipulate an expression matrix containing information on its genes and cells and many related data structures.</p>
<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span></dt>
<dd><p>Overloaded function.</p>
<ol class="arabic simple">
<li>__init__(self: ExpressionMatrix2.ExpressionMatrix, directoryName: str, parameters: ChanZuckerberg::ExpressionMatrix2::ExpressionMatrixCreationParameters) -&gt; None</li>
</ol>
<p>Construct a new, empty ExpressionMatrix.</p>
<ol class="arabic simple" start="2">
<li>__init__(self: ExpressionMatrix2.ExpressionMatrix, directoryName: str, allowReadOnly: bool=False) -&gt; None</li>
</ol>
<p>Access an existing ExpressionMatrix.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.addCell">
<code class="descname">addCell</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>jsonString: str</em><span class="sig-paren">)</span> &rarr; int</dt>
<dd><p>Add a cell described by a JSON string.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.addCellMetaData">
<code class="descname">addCellMetaData</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>cellMetaDataFileName: str</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Add meta data for existing cells from an input csv file.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.addCells">
<code class="descname">addCells</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>expressionCountsFileName: str</em>, <em>expressionCountsFileSeparators: str='</em>, <em>'</em>, <em>cellMetaDataFileName: str</em>, <em>cellMetaDataFileSeparators: str='</em>, <em>'</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Add cells and their meta data from input files in delimited format.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.addCellsFromBioHub1">
<code class="descname">addCellsFromBioHub1</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>expressionCountsFileName: str</em>, <em>initialMetaDataCount: int</em>, <em>finalMetaDataCount: int</em>, <em>plateMetaDataFileName: str</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Add cells in the format used by the BioHub pipeline, July 2017.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.addCellsFromBioHub2">
<code class="descname">addCellsFromBioHub2</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>plateFileName: str</em>, <em>totalExpressionCountThreshold: float</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Add cells in the format used by the BioHub pipeline, September 2017.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.addCellsFromHdf5">
<code class="descname">addCellsFromHdf5</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix, fileName: str, cellNamePrefix: str, cellMetaDataArgument: List[Tuple[str, str]], totalExpressionCountThreshold: float</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Add cells from an input file in hdf5 format.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.addGene">
<code class="descname">addGene</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>geneName: str</em><span class="sig-paren">)</span> &rarr; bool</dt>
<dd><p>Add a new gene with a given name.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.analyzeLsh">
<code class="descname">analyzeLsh</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>arg0: str</em>, <em>arg1: str</em>, <em>arg2: int</em>, <em>arg3: int</em>, <em>arg4: float</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>For debugging/testing use only.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.analyzeSimilarPairs">
<code class="descname">analyzeSimilarPairs</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>arg0: str</em>, <em>arg1: float</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>For debugging/testing use only.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.cellCount">
<code class="descname">cellCount</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em><span class="sig-paren">)</span> &rarr; int</dt>
<dd><p>Return the total number of cells.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.cellIdFromString">
<code class="descname">cellIdFromString</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>cellString: str</em><span class="sig-paren">)</span> &rarr; int</dt>
<dd><p>Return the cellId corresponding to a given cell name (or cell id string).</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.computeApproximateLshCellSimilarity">
<code class="descname">computeApproximateLshCellSimilarity</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>arg0: int</em>, <em>arg1: int</em>, <em>arg2: int</em>, <em>arg3: int</em>, <em>arg4: int</em><span class="sig-paren">)</span> &rarr; float</dt>
<dd><p>For debugging/testing use only.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.computeCellGraphLayout">
<code class="descname">computeCellGraphLayout</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>graphName: str</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Compute the two-dimensional layout of a cell graph to be used for visualization.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.computeCellSimilarity">
<code class="descname">computeCellSimilarity</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>cellId0: int</em>, <em>cellId1: int</em><span class="sig-paren">)</span> &rarr; float</dt>
<dd><p>Compute the similarity between two cells, taking into account all genes.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.createCellGraph">
<code class="descname">createCellGraph</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>graphName: str</em>, <em>cellSetName: str='AllCells'</em>, <em>similarPairsName: str</em>, <em>similarityThreshold: float</em>, <em>k: int</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Create a new cell graph.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.createCellSetDifference">
<code class="descname">createCellSetDifference</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>inputSetName0: str</em>, <em>inputSetName1: str</em>, <em>outputSetName: str</em><span class="sig-paren">)</span> &rarr; bool</dt>
<dd><p>Create a new cell set as the set difference of two existing cell sets.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.createCellSetIntersection">
<code class="descname">createCellSetIntersection</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>inputSetsNames: str</em>, <em>outputSetName: str</em><span class="sig-paren">)</span> &rarr; bool</dt>
<dd><p>Create a new cell set as the intersection of two or more existing cell sets.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.createCellSetUnion">
<code class="descname">createCellSetUnion</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>inputSetsNames: str</em>, <em>outputSetName: str</em><span class="sig-paren">)</span> &rarr; bool</dt>
<dd><p>Create a new cell set as the union of two or more existing cell sets.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.createCellSetUsingMetaData">
<code class="descname">createCellSetUsingMetaData</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>cellSetName: str</em>, <em>metaDataFieldName: str</em>, <em>matchString: str</em>, <em>useRegex: bool</em><span class="sig-paren">)</span> &rarr; bool</dt>
<dd><p>Create a new cell set using cell meta data.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.createCellSetUsingNumericMetaDataBetween">
<code class="descname">createCellSetUsingNumericMetaDataBetween</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>cellSetName: str</em>, <em>metaDataFieldName: str</em>, <em>lowerBound: float</em>, <em>upperBound: float</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Create a new cell set consisting of cells for which a specified meta data field is numeric and in a specified interval.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.createCellSetUsingNumericMetaDataGreaterThan">
<code class="descname">createCellSetUsingNumericMetaDataGreaterThan</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>cellSetName: str</em>, <em>metaDataFieldName: str</em>, <em>lowerBound: float</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Create a new cell set consisting of cells for which a specified meta data field is numeric and greater than a specified value.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.createCellSetUsingNumericMetaDataLessThan">
<code class="descname">createCellSetUsingNumericMetaDataLessThan</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>cellSetName: str</em>, <em>metaDataFieldName: str</em>, <em>upperBound: float</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Create a new cell set consisting of cells for which a specified meta data field is numeric and less than a specified value.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.createClusterGraph">
<code class="descname">createClusterGraph</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>cellGraphName: str</em>, <em>clusterGraphCreationParameters: ChanZuckerberg::ExpressionMatrix2::ClusterGraphCreationParameters</em>, <em>clusterGraphName: str</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Create a new cluster graph by running clustering on an existing cell graph.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.createGeneSetDifference">
<code class="descname">createGeneSetDifference</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>inputSetName0: str</em>, <em>inputSetName1: str</em>, <em>oututSetName: str</em><span class="sig-paren">)</span> &rarr; bool</dt>
<dd><p>Create a new gene set as the set difference of two existing gene sets.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.createGeneSetIntersection">
<code class="descname">createGeneSetIntersection</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>inputSetsNames: str</em>, <em>outputSetName: str</em><span class="sig-paren">)</span> &rarr; bool</dt>
<dd><p>Create a new gene set as the intersection of two or more existing gene sets.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.createGeneSetUnion">
<code class="descname">createGeneSetUnion</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>inputSetsNames: str</em>, <em>outputSetName: str</em><span class="sig-paren">)</span> &rarr; bool</dt>
<dd><p>Create a new gene set as the union of two or more existing gene sets.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.createGeneSetUsingInformationContent">
<code class="descname">createGeneSetUsingInformationContent</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>existingGeneSetName: str</em>, <em>cellSetName: str</em>, <em>normalizationMethod: ChanZuckerberg::ExpressionMatrix2::NormalizationMethod</em>, <em>geneInformationContentThreshold: float</em>, <em>newGeneSetName: str</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Create a gene set using an information content criterion.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.createMetaDataFromClusterGraph">
<code class="descname">createMetaDataFromClusterGraph</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>clusterGraphName: str</em>, <em>metaDataName: str</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Create meta data from the cluster ids stored in a ClusterGraph.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.explore">
<code class="descname">explore</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>serverParameters: ChanZuckerberg::ExpressionMatrix2::ServerParameters</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Run the http server.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.findSimilarPairs0">
<code class="descname">findSimilarPairs0</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>geneSetName: str</em>, <em>cellSetName: str</em>, <em>similarPairsName: str</em>, <em>k: int</em>, <em>similarityThreshold: float</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Find pairs of similar cells using direct computation.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.findSimilarPairs1">
<code class="descname">findSimilarPairs1</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>arg0: str</em>, <em>arg1: str</em>, <em>arg2: str</em>, <em>arg3: int</em>, <em>arg4: float</em>, <em>arg5: int</em>, <em>arg6: int</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Obsolete. Use findSimilarPairs3 instead.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.findSimilarPairs2">
<code class="descname">findSimilarPairs2</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>arg0: str</em>, <em>arg1: str</em>, <em>arg2: int</em>, <em>arg3: float</em>, <em>arg4: int</em>, <em>arg5: int</em>, <em>arg6: int</em>, <em>arg7: float</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Prototype code, not ready for production use.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.findSimilarPairs3">
<code class="descname">findSimilarPairs3</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>geneSetName: str='AllGenes'</em>, <em>cellSetName: str='AllCells'</em>, <em>similarPairsName: str</em>, <em>k: int=20</em>, <em>similarityThreshold: float=0.5</em>, <em>lshCount: int=1024</em>, <em>seed: int=231</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Find pairs of similar cells using Locality Sensitive Hashing, but still looping over all possible cell pairs.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.geneCount">
<code class="descname">geneCount</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em><span class="sig-paren">)</span> &rarr; int</dt>
<dd><p>Return the total number of genes.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.geneIdFromName">
<code class="descname">geneIdFromName</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>geneName: str</em><span class="sig-paren">)</span> &rarr; int</dt>
<dd><p>Return the numeric id of a gene given its name.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.geneName">
<code class="descname">geneName</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>geneId: int</em><span class="sig-paren">)</span> &rarr; str</dt>
<dd><p>Return the name of a gene given its numeric id.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getCellExpressionCount">
<code class="descname">getCellExpressionCount</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>cellId: int</em>, <em>geneId: int</em><span class="sig-paren">)</span> &rarr; float</dt>
<dd><p>Get the expression count for a given cell id and gene id.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getCellExpressionCountFromGeneName">
<code class="descname">getCellExpressionCountFromGeneName</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>cellId: int</em>, <em>geneName: str</em><span class="sig-paren">)</span> &rarr; float</dt>
<dd><p>Get the expression count for a given cell id and gene name.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getCellExpressionCounts">
<code class="descname">getCellExpressionCounts</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>cellId: int</em><span class="sig-paren">)</span> &rarr; List[Tuple[int, float]]</dt>
<dd><p>Get all the expression counts for a given cell id.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getCellGraphEdges">
<code class="descname">getCellGraphEdges</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>graphName: str</em><span class="sig-paren">)</span> &rarr; List[Tuple[int, int]]</dt>
<dd><p>Get the pairs of cell ids corresponding to the edges of a cell graph.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getCellGraphNames">
<code class="descname">getCellGraphNames</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em><span class="sig-paren">)</span> &rarr; List[str]</dt>
<dd><p>Get the names of all currently defined cell graphs.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getCellGraphVertices">
<code class="descname">getCellGraphVertices</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>graphName: str</em><span class="sig-paren">)</span> &rarr; List[ChanZuckerberg::ExpressionMatrix2::CellGraphVertexInfo]</dt>
<dd><p>Get the cell ids corresponding to the vertices of a cell graph.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getCellMetaData">
<code class="descname">getCellMetaData</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>cellId: int</em><span class="sig-paren">)</span> &rarr; List[Tuple[str, str]]</dt>
<dd><p>Get all meta data values for a given cell id.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getCellMetaDataValue">
<code class="descname">getCellMetaDataValue</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>cellId: int</em>, <em>name: str</em><span class="sig-paren">)</span> &rarr; str</dt>
<dd><p>Get a single meta data value for a given cell id.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getCellSet">
<code class="descname">getCellSet</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>cellSetName: str</em><span class="sig-paren">)</span> &rarr; List[int]</dt>
<dd><p>Get the ids of the cells in an existing cell set.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getCellSetNames">
<code class="descname">getCellSetNames</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em><span class="sig-paren">)</span> &rarr; List[str]</dt>
<dd><p>Get the names of all existing cell sets.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getCellsExpressionCount">
<code class="descname">getCellsExpressionCount</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix, cellIds: List[int], geneId: int</em><span class="sig-paren">)</span> &rarr; List[float]</dt>
<dd><p>Get the expression counts for a given gene id and for a specified list of cell ids.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getCellsExpressionCountFromGeneName">
<code class="descname">getCellsExpressionCountFromGeneName</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix, cellIds: List[int], geneName: str</em><span class="sig-paren">)</span> &rarr; List[float]</dt>
<dd><p>Get the expression counts for a given gene name and for a specified list of cell ids.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getCellsExpressionCounts">
<code class="descname">getCellsExpressionCounts</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix, cellIds: List[int]</em><span class="sig-paren">)</span> &rarr; List[List[Tuple[int, float]]]</dt>
<dd><p>Get all the expression counts for a specified list of cell ids.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getCellsExpressionCountsForGenes">
<code class="descname">getCellsExpressionCountsForGenes</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix, cellIds: List[int], geneIds: List[int]</em><span class="sig-paren">)</span> &rarr; List[List[Tuple[int, float]]]</dt>
<dd><p>Get all the expression counts for a specified list of cell ids and gene ids.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getCellsMetaData">
<code class="descname">getCellsMetaData</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix, cellIds: List[int]</em><span class="sig-paren">)</span> &rarr; List[List[Tuple[str, str]]]</dt>
<dd><p>Get all meta data fields for a given list of cell ids.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getClusterAverageExpression">
<code class="descname">getClusterAverageExpression</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>clusterGraphName: str</em>, <em>clusterId: int</em><span class="sig-paren">)</span> &rarr; List[float]</dt>
<dd><p>Get L2-normalized average gene expression for a cluster.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getClusterCells">
<code class="descname">getClusterCells</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>clusterGraphName: str</em>, <em>clusterId: int</em><span class="sig-paren">)</span> &rarr; List[int]</dt>
<dd><p>Get the cell ids of a cluster.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getClusterGraphGenes">
<code class="descname">getClusterGraphGenes</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>clusterGraphName: str</em><span class="sig-paren">)</span> &rarr; List[int]</dt>
<dd><p>Get the gene ids of the genes used by a cluster graph.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.getClusterGraphVertices">
<code class="descname">getClusterGraphVertices</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>clusterGraphName: str</em><span class="sig-paren">)</span> &rarr; List[int]</dt>
<dd><p>Get the cluster ids for the vertices of a cluster graph.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.removeCellSet">
<code class="descname">removeCellSet</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>cellSetName: str</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>Remove an existing cell set.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.testExpressionMatrixSubset">
<code class="descname">testExpressionMatrixSubset</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>arg0: int</em>, <em>arg1: int</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>For debugging/testing use only.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.writeLshSimilarityComparison">
<code class="descname">writeLshSimilarityComparison</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>arg0: int</em>, <em>arg1: int</em>, <em>arg2: int</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>For debugging/testing use only.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.writeLshSimilarityComparisonSlow">
<code class="descname">writeLshSimilarityComparisonSlow</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>arg0: int</em>, <em>arg1: int</em>, <em>arg2: int</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>For debugging/testing use only.</p>
</dd></dl>

<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrix.writeSimilarPairs">
<code class="descname">writeSimilarPairs</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrix</em>, <em>arg0: str</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd><p>For debugging/testing use only.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="ExpressionMatrix2.ExpressionMatrixCreationParameters">
<em class="property">class </em><code class="descclassname">ExpressionMatrix2.</code><code class="descname">ExpressionMatrixCreationParameters</code></dt>
<dd><p>Class used to store creation parameters for a new expression matrix.</p>
<dl class="method">
<dt id="ExpressionMatrix2.ExpressionMatrixCreationParameters.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ExpressionMatrixCreationParameters</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="ExpressionMatrix2.ExpressionMatrixCreationParameters.geneCapacity">
<code class="descname">geneCapacity</code></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="ExpressionMatrix2.ExpressionMatrixCreationParameters.cellCapacity">
<code class="descname">cellCapacity</code></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="ExpressionMatrix2.ExpressionMatrixCreationParameters.cellMetaDataNameCapacity">
<code class="descname">cellMetaDataNameCapacity</code></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="ExpressionMatrix2.ExpressionMatrixCreationParameters.cellMetaDataValueCapacity">
<code class="descname">cellMetaDataValueCapacity</code></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="normalizationmethod">
<h2>NormalizationMethod</h2>
<dl class="class">
<dt id="ExpressionMatrix2.NormalizationMethod">
<em class="property">class </em><code class="descclassname">ExpressionMatrix2.</code><code class="descname">NormalizationMethod</code></dt>
<dd><p>Various ways to normalize gene expressions.</p>
</dd></dl>

</div>
<div class="section" id="serverparameters">
<h2>ServerParameters</h2>
<dl class="class">
<dt id="ExpressionMatrix2.ServerParameters">
<em class="property">class </em><code class="descclassname">ExpressionMatrix2.</code><code class="descname">ServerParameters</code></dt>
<dd><dl class="method">
<dt id="ExpressionMatrix2.ServerParameters.__init__">
<code class="descname">__init__</code><span class="sig-paren">(</span><em>self: ExpressionMatrix2.ServerParameters</em><span class="sig-paren">)</span> &rarr; None</dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="ExpressionMatrix2.ServerParameters.port">
<code class="descname">port</code></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="ExpressionMatrix2.ServerParameters.docDirectory">
<code class="descname">docDirectory</code></dt>
<dd></dd></dl>

</dd></dl>

</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables</h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span>Index</span></a></li>
<li><a class="reference internal" href="search.html"><span>Search Page</span></a></li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">ExpressionMatrix2 Python API</a><ul>
<li><a class="reference internal" href="#cellgraphvertexinfo">CellGraphVertexInfo</a></li>
<li><a class="reference internal" href="#clustergraphcreationparameters">ClusterGraphCreationParameters</a></li>
<li><a class="reference internal" href="#expressionmatrix">ExpressionMatrix</a></li>
<li><a class="reference internal" href="#normalizationmethod">NormalizationMethod</a></li>
<li><a class="reference internal" href="#serverparameters">ServerParameters</a></li>
</ul>
</li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>

<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">ExpressionMatrix2  documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
    </div>
  </body>
</html>